<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Guillaume Guénard" />

<meta name="date" content="2024-10-23" />

<title>A phylogenetic modelling tutorial using Phylogenetic Eigenvector Maps (PEM) as implemented in R package MPSEM (0.4-4)</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
margin-bottom: 0em;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">A phylogenetic modelling tutorial using
Phylogenetic Eigenvector Maps (PEM) as implemented in R package MPSEM
(0.4-4)</h1>
<h4 class="author">Guillaume Guénard</h4>
<h4 class="date">2024-10-23</h4>


<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction">Introduction</a></li>
<li><a href="#preparing-the-data" id="toc-preparing-the-data">Preparing
the data</a></li>
<li><a href="#calculating-pem" id="toc-calculating-pem">Calculating
<strong>PEM</strong></a>
<ul>
<li><a href="#edge-weighting-function" id="toc-edge-weighting-function">Edge weighting function</a></li>
<li><a href="#phylogenetic-graph" id="toc-phylogenetic-graph">Phylogenetic graph</a></li>
<li><a href="#building-the-eigenvector-map" id="toc-building-the-eigenvector-map">Building the eigenvector
map</a></li>
<li><a href="#estimate-weighting-parameters-empirically" id="toc-estimate-weighting-parameters-empirically">Estimate weighting
parameters empirically</a></li>
<li><a href="#phylogenetic-modelling" id="toc-phylogenetic-modelling">Phylogenetic modelling</a></li>
</ul></li>
<li><a href="#cross-validating-pem-predictions" id="toc-cross-validating-pem-predictions">Cross-validating
<strong>PEM</strong> predictions</a></li>
<li><a href="#other-utility-functions" id="toc-other-utility-functions">Other utility functions</a>
<ul>
<li><a href="#influence-matrix" id="toc-influence-matrix">Influence
matrix</a></li>
<li><a href="#update-and-forced-pem-parameters" id="toc-update-and-forced-pem-parameters">Update and forced
<strong>PEM</strong> parameters</a></li>
<li><a href="#pem-scores" id="toc-pem-scores"><strong>PEM</strong>
scores</a></li>
<li><a href="#miscellaneous" id="toc-miscellaneous">Miscellaneous</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Phylogenetic Eigenvector Maps (<strong>PEM</strong>) is a method to
perform phylogenetic modelling. Phylogenetic modelling consists in
modelling trait evolution and predicting trait values using phylogeny as
an explanatory factor <span class="citation">(Guénard, Legendre, and
Peres-Neto 2013)</span>. Phylogenetic modelling allows one to predict
trait values when it is difficult or impractical to obtain them, for
instance when species are rare, extinct, or when information is needed
for several species and trait values are only available for a relatively
small number of them <span class="citation">(Guénard et al. 2011,
2014)</span>.</p>
<p>To apply phylogenetic modelling, one needs to have a set of species
with known phylogeny and trait values (hereafter referred to as the
<em>model species</em>) as well as to know the locations, with respect
to the phylogeny of the models species, of the species for which trait
values are being predicted (hereafter referred to as the <em>target
species</em>). Phylogenetic modelling can be performed jointly with
trait correlation modelling: it is possible to use other traits with
known (or estimable) values for the target species to help predict a
trait of interest. Phylogenetic trees being acyclic graphs, I will
hereby prefer terms belonging to the graph theory over terms
phylogeneticists may be more familiar with. Therefore I will use
<em>edge</em> over <em>branches</em> and <em>vertex</em> over
<em>root</em>, <em>node</em> or <em>tip</em>; safe in cases where I want
to be specific about what a vertex represents.</p>
<blockquote>
<h3 id="the-phylogenetic-eigenvector-maps-pem-expression">The
Phylogenetic eigenvector maps (<strong>PEM</strong>) expression</h3>
<p>Statistical maps are a type of geographic map representing the values
or states of a variable across space &lt;<a href="https://encyclopedia2.thefreedictionary.com/statistical+map" class="uri">https://encyclopedia2.thefreedictionary.com/statistical+map</a>).</p>
<p>In a paper entitled “The interpretation of statistical maps”, <span class="citation">(Moran 1948)</span> described tests of significance of
the spatial relationships among values of qualitative variables on
statistical maps.</p>
<p>“Moran’s eigenvector maps” (<strong>MEM</strong>), an expression
coined by <span class="citation">Dray, Legendre, and Peres-Neto
(2006)</span>, describes the variation of spatial eigenvectors whose
eigenvalues are proportional to Moran’s I spatial autocorrelation
statistics <span class="citation">(Moran 1950)</span> of the
corresponding eigenvectors. Spatial eigenvectors are mathematical
constructs that describe the variation of quantities across space (or
time) at different spatial scales. They were originally called Principal
Coordinates of Neighbour Matrices by <span class="citation">Borcard and
Legendre (2002)</span>.</p>
<p>Phylogenetic eigenvector maps (<strong>PEM</strong>) <span class="citation">(Diniz-Filho et al. 2012; Guénard, Legendre, and
Peres-Neto 2013)</span> are sets of eigenfunctions describing the
structure of a phylogenetic graph, which represents either a Darwinian
phylogenetic tree or a reticulated tree, i.e., a phylogenetic tree with
reticulations. The various eigenvectors describe the variation across a
phylogeny at different phylogenetic scales.</p>
<p>Contrary to MEM, the eigenvalues in PEM are not proportional to
Moran’s I autocorrelation coefficients of the corresponding
eigenvectors.</p>
</blockquote>
<p>The <strong>PEM</strong> work flow consists in</p>
<ol style="list-style-type: decimal">
<li><p>calculating the influence matrix of the graph,</p></li>
<li><p>specifying a model of trait evolution along the edges of the
phylogenetic tree,</p></li>
<li><p>calculating the left eigenvectors of the weighted and centred
influence matrix and</p></li>
<li><p>use these eigenvectors as descriptors <span class="citation">(Guénard, Legendre, and Peres-Neto
2013)</span>.</p></li>
</ol>
<p>An <strong>R</strong> language implementation of that approach is
found in package <strong>MPSEM</strong>. <strong>MPSEM</strong> was
developed to make the aforementioned process as seamless as possible. It
is a work in progress; I welcome anyone to provide relevant suggestions
and constructive remarks aimed at making <strong>MPSEM</strong> a
better, more efficient and user-friendly, interface to phylogenetic
modelling.</p>
<p>Assuming package <strong>MPSEM</strong> is installed, the first step
to calculate a <strong>PEM</strong> is to load package MPSEM, which
depends on packages <strong>ape</strong> and <strong>MASS</strong>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(MPSEM)</span></code></pre></div>
<pre><code>## Le chargement a nécessité le package : ape</code></pre>
<pre><code>## Le chargement a nécessité le package : magrittr</code></pre>
</div>
<div id="preparing-the-data" class="section level1">
<h1>Preparing the data</h1>
<p>For the present tutorial, we will use the data set
<strong>perissodactyla</strong> from <strong>R</strong> package
<strong>caper</strong>. These data from <span class="citation">Purvis
and Rambaut (1995)</span> are loaded into your <strong>R</strong>
workspace as follows:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">data</span>(perissodactyla,<span class="at">package=</span><span class="st">&quot;caper&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAGACAMAAABFpiBcAAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAARJElEQVR4nO2diZLjKBAF0f//9MS4dRQIoRteocyIbbdtwFSRq8Oq1oQBQJjQegIAJRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQTsh/Gg9i+fpMKRv0utC9hrX5+h1IXuN63NMC/l/Px/+ns0PQ7z3/3sWhr8Xw/ywtJzeE0BkGnCXybPfr4mgk6djy/HFsLz41yQMi9GDihoas4DDhC3Gt/9+rLagdoT0vXHjOT8JSlrozAQOsbVgRtDVFnSIFN1y2G41dbTQmQkcYkdQcwwaQqKrbRkJOjWJd/Ea6MwEDrEpaLyfX855tk+S0pb2lOnVGM6gMxM4xJEFu7Co8UZWCMU5QYGXBE02sjpITgq2+dqCfS1e93xtwb4Wr3u+tmBfi9cP5S/kP8PX4vXDxsp8bcG+Fq8fEPTH1+L1w0lBj+//fS25r9l+ibOCXhjYw+J7mOM3qSCoB3zN9ktcFHS/YHm+1h7Gn0n5spYTUpMBw4Vj0OFowbKtXRqbhLmllhJas4GFLUHvFyzPz+YtaFx893god9CaDSycXJlTBcu2cdxS7kqA2HRg5pqgBwuWQ/RbYrQUejOCP84Kerhg2Z45hSE+q9KrutOaTWs2D/BacCOMez2lnJCaTHOUstFEULag2ihlQ2kuDSENFqVsKM2lIaTBopQNpbk0hDRYlLKhNJeGkAaLUjaU5tIQ0mBRysaN70EfHbgxvmb7NkrZuHgl6XpLpeAXNGfVCqVsVBdUE1+zfRulbNy+Fr9dDxriO9duVocqXP9UWpL2KGXjwjHocKweNEQv2ivxc4ew3b0ySkvSHqVsXC+3y21B05bB3Ks2I2ihe2WUlqQ9Stm4VluS3YIOkWNJZd0BQVsqqrQk7XGcjfUx6LEb2O4JuupeGcdL8gKOs5FsRsOxetBh9nLa4adnR5wkKdFJNjoJ40dPsdynk2x0EsaPnmK5D9mQgyWxkA05WBIL2ZCDJbGQDTlYEgvZkIMlsZANOVgSi+NsXPhberW/MM7iYIoVcZyN81OfL21eHqEGmrNqheNsXBNUHx+zrIXjbKyLRaJ6ufU/Jjv2mMqXo4Jlobplx0vyAo6zcbxgeRhfnMQcFgWjlhp1y46X5AXO1mAqMUbw96NYcTw2tz2mX0LUpDhKJRDU4jgbRrdjBcslQbc2wOziG+M4G+tj0OINbEuCpv2yo9TC8ZK8gONsJPv5sF2wPL84N57725Ok+ByLkyQJOslGJ2H86CmW+3SSjU7C+NFTLPchG3KwJBayIcetJan2HV81nkorPMU9QZ+ahQrdBeQfBLV0F5B/ENTSXUD+QVCL44CuHkOrh4ygFscBXZ26esgIanEcEII+3VkRxwGti0Xii+mmmmk8GBh/Dvahs38KUSuUB3Ac0GhWttxuXYc01YrMRcmrYjsVENSiFNC1awx5M9OtYpjHX1WHqm1CEdSiFNDJuRhBdwqWh2iLGgvapuazBIJalAK6JuiBguXYRbuLn/0WAkEtSgGdFTTez4fCHZZzJ0lmFy+VBQSNUAroxlyUwrgLglqUAkLQHwhqUQpIaS4NQVCLUkBKc2kIglqUAlKaS0MQ1KIUkNJcGoKgFqWAlObSEAS1KAV043vQA23Pz6cRCGpRCujilaSDTZUiLYKgllcDulb9cXj0V5q2B0Et7wpapXnYrQe1hUvzpc58SwEQ1OJZ0HC0HtS0nKpEGt+ktgSCWjwLah72bmD793OqGbH9xEqZBgSNcSDogYLlvXpQo+SQCiqnKIJaPAi60/xsPah51uwmtSUQ1OJZ0MP1oPFJknnGSZI8jgW98UnSIKgFQeVAUEsngvYEgloQVA4EtSCoHAhqkRL0kUv07kFQi5Kgr47iBwS1IKgcCGpxLGiv+38EtXgW9JHR9UBQC4LKgaCWDgQ9UrCc/kuxYX5R75I8glo8C3q8YDlqMixNxtKmYNq1B0EtngU1D+WC5czWdSWozhkXglo8CHq7YDnfZBwh/nMQARDU4kDQvdGPFCxvNYm2w2xBBdncPD1ChamPv++f82w1WR5EVhdBq1EvWwc+KW2SPZ1SAEGrIS1osgXVAUGrQbaugKDVIFtXQNBqkK0rIGg1yNYVELQaZOsKCFoNsnUFBK1GvS/qr89j9fVosA8tQNBq1LnUea/l+uvR3ENNELQaHQjaAAStRu1ikWFVlDy1DOaeyvM/Jhuif1p26hAJ2mJPj6DVeP8YdNiopYuvsI/tZnmjliHpwC7+OzyTrZ1aqbyZxYJlO7nZ02lEBP0ODwlafj27BR0iRY8Imhu28OEvgqDVqCLogYJlsxtf3aN+LXY0OoL2zMuCHi5YNidCyxZ0Pkmau0/nSiF9qAuCVuNdQZ//JA0QtBoIegUErUY9QXsCQauBoFdA0Gog6BUQtBpn/kK5QOswKoOgzvhazhHUGV/LOYI648gX9T2BoM7Yu9T59ufUBkGdUUtQFRDUGceLRaJyu9wdlpfr7WN/UyUic7SAoM4oHYMOG+V2STVT9J+tusuUMTUHQZ2x8/1o3sxVwfL033JyhaDwKkbQnYLl+WbKtt/Yetm7q6wtgnbC+hh0s2A5elyOOgdTJKqztgjaCcl+PhTusGwOVwfznjVY50tVBO2Q8rqUNpBSG88fCNohO+tS2jrKbDknEBSkQVCQBkFBGgQFaRAUpEHQXnjuq0upZUXQfsgux4E1SptIfRmKoP3w1HJILSuC9sNYFWJvRDvfKSxz46W5m2kyX4sP0WDZfnVA0H6whUnzxfakGGRVHTq9a+tFli7FflVA0H6YN3qrus64cCnzT3Nn6vPmbpv9qoCgzTl/74aNgmUr2noLGj2PVu6ooE0URdDmPJbFyamw/MHRStBcuWciaHYX36xMFEGb86ig9o/mhvnL0TA/W5/sxMcEpsnSMtuvEgjanDey2M/KIGhzELQEgjaHLJZA0OaQxRII2hyyWAJBm0MWSyBoc8hiCQRtDlksgaDNeSiLFwuWd7s0XmQEbc6zV5LOjr56P33BXuFssOAI2px3Bb3diS3o13n8WnwYhuMFy8l9bEvly03q7RC0Oc9vQc8ULEf3Ect0iG7GiKBf5Pkt6JmCZbtZ3eqAoJ+mbcHyarOKoBDz6BZ03CsfL1jO/jFS3GFAUBUe2pid3vY9NPsxhOFMwXJyPrTZwR4F1ARBLU0CeuND+1kZBLUgqBwIaulG0H5AUAuCyoGgFgSVA0EtCCoHgloQVA4EtSCoHAhq8SzoC1+jK6wvglo8C7ox0PHR1y0V1hdBLR0KKjXgBRDU0oGg07X0cLJgeVWpHNJn9jJ/Osp7IKilA0H/frlSsByWZyEdxRQ8ZUd5DwS1OBC0UBM1b0GvFCxPFg+zffEmNzXz2UKsbRDU4kHQ8lurbV9G0EjRZAs6dh9sB/OQeFoBBLX0Iei5guWo5VrJxUxzDFrxfssIavEv6IWCZdsyPgMyz8xGNTfKeyCoxbegl8Y8GXL1DCGoBUEfbn4fBLV0I2g/IKgFQeVAUAuCyoGgljaCnqPFFBuCoBYHATmY4qMgqMVBQA6m+CgIanEQ0OYUO93/I6jFQUB7V5JOtD/YYLNllWwhqMVBQGcFvTPgI+PfBEEtDgI6cC0+mBKQ/YLluaHpt65UXrrFQ79+SR5BLQ4COrIFPVOwbCqdhjC1zHdY6kXsqOU53QZBLQ4CerpgeXovqXuKXzT9o/8PhnSw50FQi++AzBbteMHyEMmXCLo0Cul75jm7+Gr4DmiS6FTBcrw3zwgavxcSQdnFV8V3QNMufjhTsDwk50NmqOQkaW4ZD81JUkW6CaibQBA0opuAugkEQSO6C8g/CGrpLiD/IKilu4D8c0/Q7ngqrfAULImFbMjBkljIhhwsicVzNs4eoTiJ1ck0K+E4G8VrjiHz7Fjbq00ew/GSvIDjbBSnfkZQMZxMsxKOs2Gvvg9DXHE87f2jZ/mbKcd1UMtNw5buU/nSeF0/2/KFuOA/jrNhqzyWyiPzLH5vNnHVIRLU1J2k1XtheTdt+UJc8MNBNna+wp1+T0qO/p6MP6JnSYeCoEvhJ4K2wkE2ds5tIofySiaCpuNuCzoriqCtcJCNoqBp4fHfMeYkXbw3z3YYFkHD4V38quWTOFiSijjIxuYUp3OfkFQVjw6Z97J/45kcEyynQCH3Cct72ZYP4mBJKuIgGw6m+Chfi7eMg2w4mOKjfC3eMg6y4WCKj/K1eMs4yIaDKT7K1+It82o2rtaoZr/v/Axfi7fMu4IKjeKHr8VbBkHl+Fq8ZRBUjq/FW8axoFePT9UFUJ9fXRwLGl1DvzTegX4NbEFQSxeCPjSeCOrzq0sPgkY3Pp4um8fX6e3BQP6GyVG/yvdUXocFEz0Ian8p1i+N3Q70yzWpBYJaPAha+gY/tmhqPD1d6pfSWa36TS0zTV6/p3IMglocCFocPivakBEtOTU6JWhdRRHU0pegufLl87v4rMPs4tvgWNB5Lz/u1OPToqXN8jDNarffcmhgm1QCQS2OBe0VsmZBUDnImgVB5SBrFgSVg6xZEFQOsmZBUDnImgVB5SBrFgSVg6xZHAsaXXXviO4CuoVnQf9+JBfZHx39dpOXPvlDXPgr4BO8O/W/H92tZ3cB6VJJ0BDdKTm+tG4fylXMmcHaFDMjaDWeSfXONjopThotHTIPQ+49O9WNSuWke67JkyBoNR4SdPP1uCou45SpaTpWxZwZbIgETT/vhUpRBK3G24Im27BtQaORsnJdFvR5RRG0Gq8LOmtTEvRYFfMwzP9uR2kXv/F5T4Kg1XhfUKvPbzdunw2rh7Fj5iRpvG9yPFibYmYErca7gvbK1+JtCIJe4WvxNgRBr/C1eBuCoFf4WrwNOZnqJhdM9fhavA05K+gjo7jna/E2BEGv8LV4G/KuoNH1y/eo7guCVuNlQf9+hN2G9wgbv78GglajsqDvwBa0XyoJ+kQ96GZLc900qgBNrqk+CIJWo84xaFSzcbcedHxYWk76bVSQUCzimYcE3fx+dLGjpM80wrraI/6UId99qQsxn7A1yhMgaDWeEbTQPNp4FgSNPsB0WCmaFTTzCduj3AdBq/G6oEM4IOjhetBsv2D39MknBHbxvnlf0NV+ODl3OVUPurxoRokf4iJR80EPgqDV2Dx63DqohAFBdWFlfpAGVViZH6RBFVbmB2lQhZX5QRpUYWV+kAZVWJkfpEEVVuYHaVDl7Bf1Z786dbLyTqb5Qc6uTHwl/sz4txw41PnGJyCoKpcFfWn8d4dpNj5c5aqg2zebLd16dn5xsBfvp6GT9wYzWLDd87XQt25ui6Cq3NrFZwqXdm49GzUp1TYN823FTE1U/AnxYOnczocFipytLVkLOo0xPbWCDnl9zFYymkrcPapmGpbRps9bjWK2o6fTcKUTCJIVdLCPJwTNKRoLGg9mxxzWo4RllCthQQ/s7eIPC7oqPI7fM4JO2qZjrm2PPu90WNAB814+Pnkxu/7k2bIfjvfKmcLjZbDBKLmcLKWfYIdeTelsXOe7ANQDQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQZp/FsbWGJLELLUAAAAASUVORK5CYII=" /><!-- -->
Fig. 1. The phylogenetic tree used for this example.</p>
<hr />
<p>The <strong>perissodactyla</strong> data set contains
<code>perissodactyla.tree</code>, a phylogenetic tree encompassing 18
odd-toed ungulate species (Fig. 1) and <code>perissodactyla.data</code>,
a data frame containing trait information about the species. For the
present study we will model the <span class="math inline">\(\log_{10}\)</span> gestation weight as a function
of phylogeny and <span class="math inline">\(\log_{10}\)</span> adult
female weight:</p>
<table>
<thead>
<tr class="header">
<th align="left">Binomial</th>
<th align="right">log.female.wt</th>
<th align="right">log.neonatal.wt</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Ceratotherium simum</td>
<td align="right">6.262</td>
<td align="right">4.903</td>
</tr>
<tr class="even">
<td align="left">Dicerorhinus sumatrensis</td>
<td align="right">5.910</td>
<td align="right">4.544</td>
</tr>
<tr class="odd">
<td align="left">Diceros bicornis</td>
<td align="right">6.183</td>
<td align="right">4.699</td>
</tr>
<tr class="even">
<td align="left">Equus africanus</td>
<td align="right">5.439</td>
<td align="right">4.398</td>
</tr>
<tr class="odd">
<td align="left">Equus burchelli</td>
<td align="right">5.477</td>
<td align="right">4.477</td>
</tr>
<tr class="even">
<td align="left">Equus grevyi</td>
<td align="right">5.653</td>
<td align="right">4.602</td>
</tr>
<tr class="odd">
<td align="left">Equus hemionus</td>
<td align="right">5.462</td>
<td align="right">4.398</td>
</tr>
<tr class="even">
<td align="left">Equus zebra</td>
<td align="right">5.462</td>
<td align="right">4.398</td>
</tr>
<tr class="odd">
<td align="left">Rhinoceros sondaicus</td>
<td align="right">6.153</td>
<td align="right">4.699</td>
</tr>
<tr class="even">
<td align="left">Rhinoceros unicornis</td>
<td align="right">6.237</td>
<td align="right">4.845</td>
</tr>
<tr class="odd">
<td align="left">Tapirus indicus</td>
<td align="right">5.455</td>
<td align="right">3.919</td>
</tr>
<tr class="even">
<td align="left">Tapirus pinchaque</td>
<td align="right">5.377</td>
<td align="right">3.699</td>
</tr>
<tr class="odd">
<td align="left">Tapirus terrestris</td>
<td align="right">5.332</td>
<td align="right">3.760</td>
</tr>
</tbody>
</table>
<p>Before going any further, it is important to make sure that the
species in the tree object are the same and presented in the same order
as those in the data table. Glancing at the data table, species clearly
cannot match since the latter feature information for only 13 of the 18
species in the tree. We will therefore match the tip labels of the
original tree in the data table using the binary (Latin) species names
in a character vector <code>spmatch</code>. When no matching element
from the data table is found, the value <code>NA</code> appears at the
corresponding position in <code>spmatch</code>. We can therefore use
these missing values to reference the species that can be dropped from
the tree using function <code>drop.tip()</code> from package
<strong>ape</strong> as follows:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">match</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  perissodactyla.tree<span class="sc">$</span>tip.label,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  perissodactyla.data[,<span class="dv">1</span><span class="dt">L</span>]</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>) <span class="ot">-&gt;</span> spmatch</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">drop.tip</span>(</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  perissodactyla.tree,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  perissodactyla.tree<span class="sc">$</span>tip.label[<span class="fu">is.na</span>(spmatch)]</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.tree</span></code></pre></div>
<p>Now that the data match the tree in terms of species content, we then
need to make sure that species ordering also matches as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">cbind</span>(perissodactyla.tree<span class="sc">$</span>tip.label, perissodactyla.data[,<span class="dv">1</span><span class="dt">L</span>])</span></code></pre></div>
<pre><code>##       [,1]                       [,2]                      
##  [1,] &quot;Dicerorhinus sumatrensis&quot; &quot;Ceratotherium simum&quot;     
##  [2,] &quot;Rhinoceros sondaicus&quot;     &quot;Dicerorhinus sumatrensis&quot;
##  [3,] &quot;Rhinoceros unicornis&quot;     &quot;Diceros bicornis&quot;        
##  [4,] &quot;Diceros bicornis&quot;         &quot;Equus africanus&quot;         
##  [5,] &quot;Ceratotherium simum&quot;      &quot;Equus burchelli&quot;         
##  [6,] &quot;Tapirus terrestris&quot;       &quot;Equus grevyi&quot;            
##  [7,] &quot;Tapirus pinchaque&quot;        &quot;Equus hemionus&quot;          
##  [8,] &quot;Tapirus indicus&quot;          &quot;Equus zebra&quot;             
##  [9,] &quot;Equus grevyi&quot;             &quot;Rhinoceros sondaicus&quot;    
## [10,] &quot;Equus burchelli&quot;          &quot;Rhinoceros unicornis&quot;    
## [11,] &quot;Equus zebra&quot;              &quot;Tapirus indicus&quot;         
## [12,] &quot;Equus africanus&quot;          &quot;Tapirus pinchaque&quot;       
## [13,] &quot;Equus hemionus&quot;           &quot;Tapirus terrestris&quot;</code></pre>
<p>Since they do not, we need to recalculate <code>spmatch</code> with
the new, reduced, tree and re-order the data accordingly:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">match</span>(</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>  perissodactyla.tree<span class="sc">$</span>tip.label,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  perissodactyla.data[,<span class="dv">1</span><span class="dt">L</span>]</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>) <span class="ot">-&gt;</span> spmatch</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>perissodactyla.data[spmatch,] <span class="ot">-&gt;</span> perissodactyla.data</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">all</span>(perissodactyla.tree<span class="sc">$</span>tip.label <span class="sc">==</span> perissodactyla.data[,<span class="dv">1</span><span class="dt">L</span>])</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>The last code line is just a last check to guarantee that all species
names are matching. As a last step before we are done with data
manipulation, I will put the binary names in place of the row names and
delete the table’s first row:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>perissodactyla.data[,<span class="dv">1</span><span class="dt">L</span>] <span class="ot">-&gt;</span> <span class="fu">rownames</span>(perissodactyla.data)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>perissodactyla.data[,<span class="sc">-</span><span class="dv">1</span><span class="dt">L</span>] <span class="ot">-&gt;</span> perissodactyla.data</span></code></pre></div>
<p>Our data of interest now appear as follows:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">log.female.wt</th>
<th align="right">log.neonatal.wt</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Dicerorhinus sumatrensis</td>
<td align="right">5.910</td>
<td align="right">4.544</td>
</tr>
<tr class="even">
<td align="left">Rhinoceros sondaicus</td>
<td align="right">6.153</td>
<td align="right">4.699</td>
</tr>
<tr class="odd">
<td align="left">Rhinoceros unicornis</td>
<td align="right">6.237</td>
<td align="right">4.845</td>
</tr>
<tr class="even">
<td align="left">Diceros bicornis</td>
<td align="right">6.183</td>
<td align="right">4.699</td>
</tr>
<tr class="odd">
<td align="left">Ceratotherium simum</td>
<td align="right">6.262</td>
<td align="right">4.903</td>
</tr>
<tr class="even">
<td align="left">Tapirus terrestris</td>
<td align="right">5.332</td>
<td align="right">3.760</td>
</tr>
<tr class="odd">
<td align="left">Tapirus pinchaque</td>
<td align="right">5.377</td>
<td align="right">3.699</td>
</tr>
<tr class="even">
<td align="left">Tapirus indicus</td>
<td align="right">5.455</td>
<td align="right">3.919</td>
</tr>
<tr class="odd">
<td align="left">Equus grevyi</td>
<td align="right">5.653</td>
<td align="right">4.602</td>
</tr>
<tr class="even">
<td align="left">Equus burchelli</td>
<td align="right">5.477</td>
<td align="right">4.477</td>
</tr>
<tr class="odd">
<td align="left">Equus zebra</td>
<td align="right">5.462</td>
<td align="right">4.398</td>
</tr>
<tr class="even">
<td align="left">Equus africanus</td>
<td align="right">5.439</td>
<td align="right">4.398</td>
</tr>
<tr class="odd">
<td align="left">Equus hemionus</td>
<td align="right">5.462</td>
<td align="right">4.398</td>
</tr>
</tbody>
</table>
<p>Finally, for the sake of demonstrating how to obtain predictions, we
will remove the Sumatran rhinoceros (<em>Dicerorhinus sumatrensis</em>),
the first species on top of the table) to obtain our training data set
{<code>perissodactyla.train</code>}, keep the withdrawn data as
{<code>perissodactyla.test</code>}, and calculate a tree without the
target species:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>perissodactyla.data[<span class="sc">-</span><span class="dv">1</span><span class="dt">L</span>,,drop<span class="ot">=</span><span class="cn">FALSE</span>] <span class="ot">-&gt;</span> perissodactyla.train</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>perissodactyla.data[<span class="dv">1</span><span class="dt">L</span>,,drop<span class="ot">=</span><span class="cn">FALSE</span>] <span class="ot">-&gt;</span> perissodactyla.test</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">drop.tip</span>(</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  perissodactyla.tree,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">tip =</span> <span class="st">&quot;Dicerorhinus sumatrensis&quot;</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.tree.train</span></code></pre></div>
</div>
<div id="calculating-pem" class="section level1">
<h1>Calculating <strong>PEM</strong></h1>
<div id="edge-weighting-function" class="section level2">
<h2>Edge weighting function</h2>
<p>As previously announced, I use the vocabulary of the graph theory
when describing <strong>PEM</strong>: a tree is a (directed) graph, a
branch is an edge, and the root, nodes, and tips are vertices.
<strong>PEM</strong> allows one to specify a model of trait evolution
along the edges of the tree. The trait evolution model is a power
function of the edge lengths, with parameters <span class="math inline">\(a\)</span> and <span class="math inline">\(\psi\)</span> describing the shape of the
relationship between the edge lengths and trait evolution rate:</p>
<p><span class="math display">\[
w_{a,\psi}(\phi_{j})=
\begin{cases}
\psi\phi^{\frac{1-a}{2}} &amp; \phi_{j}&gt;0\\
0 &amp; \phi_{j}=0,
\end{cases}
\]</span> where <span class="math inline">\(a\)</span> is the steepness
parameter describing how abrupt the changes in trait values occur with
time following branching, <span class="math inline">\(\psi\)</span> is
the evolution rate of the trait, and <span class="math inline">\(\phi_{j}\)</span> is the length of edge <span class="math inline">\(j\)</span> <span class="citation">(Guénard,
Legendre, and Peres-Neto 2013)</span>.</p>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAHgCAMAAABNUi8GAAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAS/klEQVR4nO2dibajKhQFzf//9Ht9YxIHQMbDBqrW6s6EgFqXScTtBSDM1jsDACEQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEhQps/5Oz2fe/cBCAcjJUQlCwYcsrQhEUTIgQzbvdg9gICuVEiObZ8P+tEBSa8yxaYNuCXwFaQhsUtHnuWyEoSIOgIA2CgjQICtIgKEiDoCBN9hyU/W3mRViAOHIuoB4FrZcTAAfphp2uaSEotKW0in9Rw0NLitugmXEARFEo6PXz9qMkVwA7lQUtixjgSoGg39odQaEZGb34vf7+/u+OAkGhBgUeMdEU2tPMIwSFGiAoSIOgIA2CgjQICtIgKEiDoCANgoI0CAqJnC8d7p/c1xPjQ/pTSwqtEDH0ZbtMCP7MyXCc8JOLwZDB5NqAoHNynh60f3LPGbrfe+EL+ZReCxB0ThIEvRa1CArtQVDoQHzPBUHBmqT7cxAUTEm+fezbDd8Onzx988O9F/TiIYOcmxu/45nb6ZMz8sO9F4yDQiKFt97Gb1ySDIKuyiB3hiPokgxi5wtBV2QcO18IuhyjLfqCoEsxmJ0vBF2J8ex8IegyjFa1f0DQJRjUzheCrsC4dr4QdHpGrdo/IOjUDG7nC0FnZvTC8w8EnZUZ7Hwh6KRMUXj+gaATMo2dLwSdj5nsfCHobEymJ4JOxXR2vhB0ImbUE0FnYU47Xwg6B9PqiaATMLGdLwQdnrn1RNCx6WLnMc39wda+51ufvmLhhtXoU3je167ZTi/HH7ZzKJa+WYheVft99S+/nywetiz9Wp43w7x+nr9E0HXo2i+6GvZdGcyVJwRdkBZ6JsToMMxfhiLoajQpPJPiRFDwUlvPrGnN32743icPaXdUl1787FS1syCy0wK22+nlGpAFbJehYuH5LfkqxFM7ZN1t+0S8IvX0HO5mJQSVp5ZSg5m5g6DiVNGzVq3eAQSVpoaew9XqJxBUl3KxRjZzB0FVKdUzdcBRFATVpEzPsWv1EwgqSIle05i5g6ByFJV+s/mJoGrk6jlRrX4CQaXIk2zmY42gOmSWgXOWnB8QVIUMPSet1U8gqASztiDLQVABku1cSGcE7U6abeuo+QZBO5Oi57aenwjalQQ7V22lImg/opVb+VgiaC/i9Vyz6NxB0D7EWSep5jlTfzvC6naTEXOSVOdznvPlfvv55vgV98UPw7Oewj2i7fvf6Y3TT1YWGRBd9eLwCercKwQdjUc95f29CvrJMIJOQNi+XmqmpXsRdG9Wek43go7EQ+Fo7eennZvTq95unxB0dAIeGFfrhT2wbzf866W/Y74dXujFC+NXwl/61E++0h+Cc3U7T7KsbjcCXjFajya1jT8+6sICuw0IutOjXy48jpoKgrbFY0p9garW4kIgaEueL1BXSCG5YzwUCNoOt56VDsxEtXgQBG3FXaByow795lVA0DY4nymQH9t7+6XM3EHQFlxMKljOxhHbWiBofe51e3oMc/d8EkDQ2pxn6SbOxFiyFg+CoHU5CpZwCDDTB4JW5GhZpHGY+QCCVuOnWsy+Y2YcCFqJr29P5mFmEghah9CND98gmJkOgtbgO1PDOy/SNj8TgaDFBGYjY2YxCFrIZ2WN65cT7/9p14Lz5ONDxiWWs413tnhGxOPxT8TLwPz0Ox6/XkinlUVcyVaJeDx+axKtYOab+PVCOq0s4kq2SsTDMesk9ifitesiqCvZShEPxTpl5pWRBd1+3De71IYjv/61PQXyUfM1mpEFLYt4AP7a+cuWnW8mFXSCc/ouONeeQPz6nsnPud9+L/khnxIrzWBExCOf1W+Z6W68LMXnEHwbCK/DS2bIUHKVMvgc8Zjn9ejjdvm8OPEHouSQFWwb3nR4QV33FaGnOc0O+MiCvovK23fo2QE7QUfh3srMazxBFRD0h6eE5N7fniDoH08DeejZC9qgL3/j8ncXh2Fm4MTaggZGNQ+jS4o5X4Z1BQ3m5ycuXfe+rCpoSLvjpVz07Mxygj5eqzxdOxLJ9MIs1YtPzBN6CrCQoM/V9flmFr09WJElBI1sSZ71VNqBhZm/DRo3AdF9dRO6M7egsbM3bzOXmuQGMphX0Oj0rgHRU4lZBY1uQ14D0vjUYkZBiyZwo6cW3vNReqK6neh4xe4h0VMO/xkpPFd9+kRFoandBXFP0fX+UhhxW8qSRE9J7ARtevpTy857cPTUxDWLfKtx97epoMVxo6cq/hK0dsSNFEhVy7PCQJ3MQHVGH2ZKjdXtZ5WsrMJlSs2+XsBzsykUMpCaJw/FhUp7QSt5hZ5pnOY27L0Vj0Xnud+H/9OSi/yyRsQ1Sa/bXRtQuydy7kCHBHWEXEbQapGjZyqOEZ73in/uKvMuaGKRMKKg6aXew23FEM9d0G0/kLfrHqeQ26cN6gj5mF7kt0m0aoPmRONbMQQ/3yQLc2xaXstIT8ijmWWCbluVgdA2gmZky7cBer6PTWpT/vvf6/zusQR9VRK0ErZXkrw8LRqyKBlmHjb9dciPxrnO+CVkrSXAM0cEwhGXGpGnlGejdf3cis/Eb7Hp37Cmu8Y9hvwESa2b/eWLb3QrN+I6h6UoCwVxTUHd3Y6PrSRdXwm6VRfUPDLv4uTr+VlQo/dmBEHrGrWUngObuTOCoBmJU7u/Rjdzx9sGLfSzVhu0auW+ip/jl5s/tIeZsip3/ybT6zmTmTuBs1k5YqMDF5Bwcj/n3DtdQTPHPRes3ScsN3+4hlfTZ5XGRJx6Sa1q5T6pn+Wj7vL4ryTVFrQ9a9XuU5ebP+yq+JRta17VzI9QlEXM3LHrxSdsSu3uZaJdiUN7mCk+tVDtPsVJXavc/BEYqK8ccWSE1O43yk/GwHg6SRUmZUWlFJefgk0G93PVcvOHmqAZCU1afC5dbv4YXtAZe0eUmz/sBH3eImOTyvH1h3Lzgq+TVH/xsBZbTFV8Um66kBkHXbp3NFS5eTyy+0Vx07WZ+tx2vGz1Ply5eW/92a/N9KpwhpMEzbjbfQo/B8nmkfs8DV9/5RxyO70kpnf/tvzIte3FD9/6LO+FduIm6LsKsF2byVjQymdK3k/1/AVxC+pqqFwFrbk2k2kVn74UWPBHZT9ly834XF0F3RzvXCH112byBkxMaNTWp0pvaPP8S9g8S9DDpxolaDF9zoWqn72yVWBiIMpDh/z21hdyD+IOGU6uDZERpxafDz/r+WlWq28tbHQmtFt3qrSN12aqQFwbNNGowar3lrW6kY0POagdsu62aRG7Ba2ZopafVXfN8W8ROgtaFGPl+KtRXKvfa+t16SpoWmNkiOo9KxvbsuXjMz0Fnc7P+KE1jIylYy++nlT9B+efavUNGzPpKWityDrr6bOOMrIGvcdB4+LSrd63y3uMrEy3NmjF+r1WTKkJY6QB+oLqVe/b/pAJrDSgl6DxfmpU7/S7O6E+UN/VT0ffGzONURc0nEaj3PvKSey0p08vvkqqlfUM195D3Xc5E10EjVTLonqP6e0UX1uHfPoIWh5FqZ+xvR16RJ3p0gatkGi2nynGYWZ/ZDtJdYc/M8YtsVOCHoLGpFmr/Zk5oo6dKnQQNCrJ4j5W9rUe+utSaApaUnyWXIWkvy6HfS++nZ/lz2fGTjn6DNQ/bJpevxePBmFnNNdFFR1fOkPuRzj1Rt6k0DYRp/hZY5wSM1M4/yn/GlT3o3j9cvOGDCfXBl8btHx8/fe+RuaxM43t+9/h0/lLZ8g9mDPkY3otyBX06aLj9xU7u5At6PYaQtBiP+uNAmFnDrmCHqv3SQVNXloyM50FSTgaswuaH1/F2XXYuZMxZpwt6P64BVfIx/RakBOxZ5t365qysyZbspm/LY+VYVQv/tJ1EO3FPyfnXMHvcyDqZBQ7y3uYp0UVY8ZBv5VfzpmUEtTx+bNr9NnLqdmGT4qq6M+hYNu0iP8OT+ZOtXiow0JUNtMUY0HzNu7wXMY5GNnMHZle/OZ49/miLJODn6EsJjBzR0XQzR+qyM85zlICqb1kdaSqeOfPBX5OdaaemM3MnYx9cjztNjJir2uB4duiu+NWYZ4a/Ub6bp3+Uv2bOwX1B/b9/ef6OenpujKxmTvJe3e5gJUUsbcA9c4AyfNz7lP2x/xm7hQL6rtG7mqDBmv4pB8CzH7aJm1r+igV9PKXvG2+R32GbiXy/pLTAJn37M28b14KBfXH4bpumZiBLD/nZJka/YaloO5mqb9kTc3bjGdwXTN3Cnrx32OXLehfDIGeU2K25mKxtqaP/HHQh4fXui4JZaQSGTgl5gGYbX8KKDgUcdcu/V8Ft8+d9zQ4y9foN+wudTp/L6/fJzmdmOnBTFD354y+/SXcDGd1hn1oRSdB90++gjLez7Gh3HzEUNAt8Gt425JQotBHj6SLoJsnzOfriEyNfHZHzrs5PQXN93PQU0yNno5hL/5WbnoGUKcsP+fozXWgo6C5fo52pik3S7AfB33oHz1kaKgzTU+oHPNhpiIBBzrdA2VVmk7DTLFbDQg1elWsBQ33kMKX54vz1J4R8jgWdoJuz4IGMqN+5ik3G2Eo6HNymR2r7ohnb2jsevHPZYz/V1kD6Ke3ptN0u5QKXtUA1XzNha2goRZo/Mz87lBu2mE6Dprhp5wJavmZHeNOkjfV+AWeukG52QNLQUNpJY46WcMoUi9EBBUuP5mH1JU+VfwtsH/uU08oN/tjOA4a6CI5B007I5AF6DIfNEbHvnZQq+vQaaD+HPQStqcd1OpiKLRBt8AnUzBTD+vZTK5fwhvaQK0uir2g4aK1hyaMwAtjX8VfC8y+FTxmimNegoZ6RNgJV3qMg/q+NfSFWn0Ueg8zdfCTkaSRsBbUW8EbKYOZg2HdSeoqKHaOR19BQzOYa2cHO4ekq6B21Tp2jkrPS53ea0uVs4GdA9NV0FZpmyUAzTHuxTtq+FY5wM4psBX04c65xqnDgBheSbp86Zk70iJpGBfbNqhBAYqdc9FN0JjFxJJTxM7p6NaLr5wwdk5KL0Grjk9i57yYCtpk6hJ2To3pMNN5akiNpFFzdjoK2iIFmA3TcdDN9b5W/DAlfTpJpYJi5zJ0EbTMT+xciU6CZieLnYthKWjx/HnsXI8eguYVoIx3LonlMFPJ/CXuKVoUQ0G/BWhGmti5KrbzQd9vUtPEzoWx78Unpoida9NB0IQksXN5DAVN7SNhJ3QQNLYARU/4h72gWVvDqlgPM+EnJCEoKHbCD+tx0Kf0sBNOGA8zPXSR0BMuWAvaJS8wLkKC4ifcsRXUnxp2ghMNQdETPNj24hEUEjG9Lz7hS4A/7Kr4l3uQCT0hhK2glhmAKTDtJPln2QO4Mb2rcwsHAbjR5b741mnDPPQTFD0hAsvpdigJyXQSFFchDsNhJvyEdOzv6myZKExHD0HxE6LpICh+QjzCCzcAyN4XD/DGcD4oakI6xrcdIymkYXzTHEAaGvckAXigBAVpek63A3gEQUEaqniQhnFQkAZBQRqqeJAmQ5uzaj7xEBRqkK7N+f52193uzogRFHJI1uZ8Sd1/gZ17jKEGCArSVBZ0+1GaEMCrYQkKUAMEBWkKevHb6RNAC/LHQbeoJx8BFFGgF2ZCe7AMpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpzATdQBorD1KxE5R0hJPRrUkRVDudyXYnHQTVTmey3UkHQbXTmWx30kFQ7XQm2510EFQ7ncl2Jx0E1U5nst1JB0G105lsd9JBUO10JtuddBBUO53Jdicd2YwB/ANBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWnM5hs2TugYv8VSGc2Pm+X+sLJI84eBnOI32KfmJ/T0B9c6LYtEcrHJV+tD4Hh4U1O21mmcEkBQm1QMBW1fZZlW8Qb7g6CWgm5tUzslZpDAFPuTy4yCvu6f2qRomUDbJrysnwhakqJlAk178Q3jLmW2Xvy3NpxEUIv9UfZztnHQd2+i+bhe84HWTwIm+yO9xLJqvnJhfyZj+QMA2iAoSIOgIA2CgjQICtIgKEiDoCANgoI0CArSIChIg6AgDYKCNAgK0iAoSIOgIA2CgjQICtIgKEiDoCANgoI0CArSIChIg6AgDYKCNAgK0iAoSIOgIA2CgjQIGkB51bdV4PiHcK7OySGzhKMdwrk4PIfMEo52iI+g74r+XeF/qv3903mNWdoEteFwhjhV8Z83lwfEvF9exxeoB0czxLkN+llmfP/tr7Q8mim+mPaYcDRDXDtJhycWncvMwyMPoCoc0hDeKn77NE0dnnJIa8LRDHB41sale7T//1OSTlIjOJwgDYKCNAgK0iAoSIOgIA2CgjQICtIgKEiDoCANgoI0CArSIChIg6AgDYKCNAgK0iAoSPMfYoXbS4Cd5pkAAAAASUVORK5CYII=" /><!-- -->
Fig. 2. Output of the edge weighting function for different sets of
parameters <span class="math inline">\(a\)</span> and <span class="math inline">\(\psi\)</span>.</p>
<hr />
<p>As the steepness parameter increases, the weight assigned to a given
edge increases more sharply with respect to the phylogenetic distance
(or evolutionary time; Fig. 2). In the context of <strong>PEM</strong>,
the edge weight represent the relative rate of evolution of the trait;
the greater the edge weight, the greater the trait change along that
edge. When <span class="math inline">\(a=0\)</span>, trait evolution is
neutral and therefore proceeds by random walk along edges. When <span class="math inline">\(a=1\)</span>, edge weights no longer increase as a
function of edge lengths. That situation corresponds to the scenario in
which trait evolution is driven by the strongest possible natural
selection: following a speciation event, trait either change abruptly
(directional selection) at the vertex or do not change at all
(stabilizing selection).</p>
<p>Also, the shape parameters may differ for different parts of the
phylogenetic tree or network. For instance, the trait may have evolved
neutrally at a steady rate from the base of a tree up to a certain node,
and then, may have been subjected to different levels of selection
pressure and various evolution rate from some of the nodes towards their
descendent tips.</p>
</div>
<div id="phylogenetic-graph" class="section level2">
<h2>Phylogenetic graph</h2>
<p>The first step to build a <strong>PEM</strong> is to convert the
phylogenetic tree. The is done by giving the tree to function
<code>Phylo2DirectedGraph()</code> as follows:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">Phylo2DirectedGraph</span>(</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  perissodactyla.tree.train</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.pgraph</span></code></pre></div>
<p>Here’s a snipet showing how the graph container used by
<strong>MPSEM</strong> stores graph information:</p>
<pre><code>## List of 2
##  $ edge  :List of 3
##   ..$         : num [1:21] 13 14 15 16 16 15 17 17 14 18 ...
##   ..$         : num [1:21] 14 15 16 1 2 17 3 4 18 19 ...
##   ..$ distance: int [1:21] 9 4 3 1 1 3 1 1 5 2 ...
##  $ vertex:List of 1
##   ..$ species: logi [1:22] TRUE TRUE TRUE TRUE TRUE TRUE ...
##  - attr(*, &quot;ev&quot;)= int [1:2] 21 22
##  - attr(*, &quot;class&quot;)= chr &quot;graph&quot;
##  - attr(*, &quot;elabel&quot;)= chr [1:21] &quot;E1&quot; &quot;E2&quot; &quot;E3&quot; &quot;E4&quot; ...
##  - attr(*, &quot;vlabel&quot;)= chr [1:22] &quot;Rhinoceros sondaicus&quot; &quot;Rhinoceros unicornis&quot; &quot;Diceros bicornis&quot; &quot;Ceratotherium simum&quot; ...</code></pre>
<p>This list contains two main elements, <code>$edge</code> and
<code>$vertex</code>, plus additional information. <code>$edge</code>
and <code>$vertex</code> elements each contain a list of
sub-elements:</p>
<ul>
<li><p>Element <code>$edge</code> is a list containing information about
the edges of the graph, namely the indices of their origin and
destination vertices (the two first unnamed elements) and an arbitrary
number of supplementary elements storing other edge properties. In the
present case, a numeric vector created by
<code>Phylo2DirectedGraph()</code> and called <code>$distance</code>
stores the phylogenetic distances (<span class="math inline">\(\phi_{j}\)</span>), which, in this example,
correspond to the branch lengths of
<code>perissodactyla.tree</code>.</p></li>
<li><p>The element <code>$vertex</code> is a list containing an
arbitrary number of elements storing vertex properties. In the present
case, a logical vector created by <code>Phylo2DirectedGraph()</code> and
called <code>$species</code> stores whether a given vertex represents a
species (<em>i.e.</em>, it is a tip) or not (<em>i.e.</em>, it is a
node).</p></li>
<li><p>In addition to edge and vertex information, the container stores
other useful information in the form of attributes:</p>
<ul>
<li><p><code>ev</code> stores the number of edges and vertices.</p></li>
<li><p><code>elabel</code> stores the edge labels.</p></li>
<li><p><code>vlabel</code> stores the vertex labels.</p></li>
</ul></li>
</ul>
</div>
<div id="building-the-eigenvector-map" class="section level2">
<h2>Building the eigenvector map</h2>
<p>In <strong>MPSEM</strong>, <strong>PEM</strong> are build using
function <code>PEM.build()</code>. As an example, let us assume that the
steepness and evolution rate are <span class="math inline">\(a=0.25\)</span> and <span class="math inline">\(\psi=2\)</span> among genus <em>Equus</em>, <span class="math inline">\(a=0.8\)</span> and <span class="math inline">\(\psi=0.5\)</span> among genus <em>Tapirus</em>,
and <span class="math inline">\(a=0\)</span> and <span class="math inline">\(\psi=1\)</span> from the root of the tree up to
the vertex where the two latter genera begin as well as among the other
genera. The following figure will help us figure out the indices of the
edges involved:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>perissodactyla.tree.train <span class="ot">-&gt;</span> tree</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">&quot;N&quot;</span>,<span class="dv">1</span><span class="dt">L</span><span class="sc">:</span>tree<span class="sc">$</span>Nnode) <span class="ot">-&gt;</span> tree<span class="sc">$</span>node.label</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">plot</span>(tree,<span class="at">show.node.label=</span><span class="cn">TRUE</span>)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="fu">edgelabels</span>(</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>  <span class="dv">1</span><span class="dt">L</span><span class="sc">:</span><span class="fu">nrow</span>(tree<span class="sc">$</span>edge),</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>  <span class="at">edge=</span><span class="dv">1</span><span class="dt">L</span><span class="sc">:</span><span class="fu">nrow</span>(tree<span class="sc">$</span>edge),</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>  <span class="at">bg=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a>  <span class="at">cex=</span><span class="fl">0.75</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAHgCAMAAABNUi8GAAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAVlUlEQVR4nO2diZakKBAA4f9/et9sKQKSisqRSUe83a6uKg4TY8ADaecBFONmbwDAFQgKqkFQUA2CgmoQFFSDoKAaBAXVICioBkFBNQgKqkFQUA2CgmoQFFSDoKAaBAXVICioBkFBNQgKqkFQUA2CgmoQFFSDoKAaBAXVICioBkFBNQgKqkFQUA2CgmoQFFSDoKAaBAXVICioBkFBNQgKqkFQUA2CgmoQFFSDoKAaBAXVICioBkFBNQgKqkFQUA2CgmoQdALuxOwt0gtNM4FTo7MXRGiaCQwVtL6LVumCyo1anX+NHqxxvrOgn1NOdQRBJ+C2dv+J6rULOhWVG7U6kaBuWA/6r8d2e72h+nj0d/73zu0vW7Ltp3MhSTF7382Hgbj4tbugLnTSLhc0bMCWMvnQxQn2DE7O3mnz+xYPJVz8MqgHLZuZdIBBwtBbngW9yN5186HA+XJlK36lb7V0rcml1Zz7vmSIP77zVYIOUBRBZbq1TdJpjj8GdS7TNU55M8QfL6fsfTcfzvQUNO23Rl0Hdbun0cupB02OWrf8x0lSXgo96Dy6ChrtdzdsLxjc2wY3eRhdh/gxNU2qpyEGN3kY6wlqEJpGpp+gwpk2nKFpZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZGgbBbATZHgmSQE0jQxPdSqAppFZQ9AXXbSmLl3Pluij79pMLlpDbMTaTM9yuOz9NBBUpvvaTC76rxvvBFWDqo1RxgBB45d+lf1eb1dYjlcN+z9RvE7k/p0bvNAygsp0b/QxgkZr1V2vsLxv0y6mPxRMUg5daBlBZXoLmnWk3SsrOJX2fPuKe0nG9Jg0Wo1vzELLCCpTv4jxU36l+0ObnmzB+EIP6hO5ouPi8DZkTZJcldIYA4JKTd6/4p4Fu/Db8GPQqxWWL3vQPF+xlE6br5jTJq4gqEt+7X0MWrfCcvgwJA7545Ok9ByLk6RFBXVhj/es6VyxNQxscjQY+u79zani6BCtWT+x3j+5fhjY5OQo3E0TtOGR1jRBDWKgaaJrbt4N7UGT0beloLNO+wxioGmS892xQ7zrIyjU86nFm12quyZs6Pb/oGr3PwcYR/ulteAN3wRttRUVtcSCDqn2uC4YbQaGjsaWoL9+zfxJEtRjStDtFUH/EEYETY5FzV8HhXqMCDqhWrpLFSDog4phPBYELVz+GVPxoHrgAgOCTmPx8GyAoDKLh2cDBJVZPDwbIKjM4uHZAEFlFg/PBggqs3h4NkBQmcXDswGCyiweng0QVGbx8GyAoDKLh2cDBJVRFN60273TQVAZReH93edAEVRGUXgfBH3b4SqJHkFlFIUXP9nqnq0N8DYKJdEjqIyi8DY1w28I2j+zfhSF91Pzk6DudoXl8Kz1/sR19DLtgRcElVEUXvRkq/f+0ZOtm1lxvou1kd3xT8Adb128FWNBUBlF4aXHoG960LKZea+495XRqp97/kldKILKDAjvwSIn8QY9WpP5COZmhWUffewzQc997SAQVGaEoI/SxYI+ruLoeaUVllMX4yE++D0eBJVRK+izy0xpN+rkFZaLJ0nREE8Pqgxlgh76PMj4rjpFIKiMMkFfZWyRayoIKrOeoAZBUBlNggrn5uuDoDKKBP27IKgMgioAQWUQVAEIKoOgCkBQGQRVAILKIKgCEFQGQRWAoDIIqgAElUFQBSCoDIIqAEFlEFQBCCoz4Eb54i3YAASVGTDVaPEWbACCyiCoAhBU5vnzGNsA/1DQ9KGL6rx/AwSVeSRo8vzOsypc+R38A0FlHj0w5N8JmjxfhKBnEFTmxRC/LxdTX4XLBGWMT0FQmWdLIwS7Hq6qEB+yznq2VzEI+h0X/fSpcTVZsz7zfaMu+eASgn7no6C+maC3HxgEQb8TH4P6r4J+aNOyoPV96VWi/Lv9uKa7Agj6nXAF1IdVY55kzY5Bv21HZOXjqwJPBM0uWPQzAUG/o+VOUnY9trT85+uNuRG0Hwj6HU2Cnpdh3qtw9yss+7jfTZdWPi20nAjacaRH0O98OHtu34OeBXW1Kyw7l3yYvMu+Y4j/I3QStHAMWjaztHxtLOgpe/j3h6B/goYtmNww+P0Sd+fFHtQnitYImm85gq5NS0GlIX7//naF5WgYd6mgJbGTABB0UUYImnaj7mKF5eNE6OhBs6MFF5LsV9WOly4g6FSaC3p9PdbgDkPQqXToQa9KN7jDEHQqgwU1CIJOpa2gzGZqmRl8Y0GXBEGngqB3IOhUrlrw4YC96M5A0KlcCvog7f3XVkHQqSDoHQg6lVtBt7tDFdOg09uWUdb3W6cBBJ3KnaDxzKGHgoastvcSgk7lRtAny0Fkf2M2mblsGASdStUQXylo3lsiqO3IVVApaOUxqEPQhpnBF6515tc9w4Ho/Xole8pQtkfQVlsBBdIh/vZp++JZPIJCN1oIylk8dMP5R8tBZIIWR32DIKhauJP0DwRVC4L+A0HVwmymfyDoKiy6MxB0FRbdGQi6CjyT1DgztIWnOhtnhrZIghrvSxF0FZKlb35SdrnROXinI+gqhKWY8g9mbVAbEHQV9hnKLrxPBXW3KyyHJC4sEOZ9XOqMW6cIugrxEO/jWSaRZzcrLCf/x3NNCusujgJBV0EU9Hhf6kGTEqJs4eQKQaEJ+2Ad/YxP4Is9qE8VdcVn9MKzJA5B4QtnQY/38TGouMLyuZyj++x2VeAGBF2HXM1iN+rkFZbjw1UffRcbPPyiKoKuQ7gCehql4xTX+aUk065YIeg63Pt4s8OuesdZt6MQdB2e+mgCBF0HZjM1zQxQAYKCahAUVPNd0PxmGUBDmgq6xFE5qOKboNF9B39MVwBoxsce1DHEQ1e+DvHnJSkBGvL5GPQ0oxCgIZzFg2oQFFTTUtBVbv+CIriTBKpBUFANgoJqEBRUg6CgGgQF1SAoqAZBzbHko0ciCGqOJR/eFEFQc7x52r3NnmLhBqjgWDUkXkj5PkfNh9dJWJsJKthWXDrevhP0bc2DQVBz/J6zcce7OkH3RZv2NRa3YsJKTu60vE2cJKxu55LCivmagqDmcNFPnz4VdpnD/2zzh1iJ7IXVFYOQiZtHXilfSxDUHLGg0VJ2dznyP/wRaXX0hOd8JwmjfxJSvpYgqE7OFzuzy55Jr3hJKtq5B03eJ7u0VtCeiiKoTi6a9ixoTWHbGB2OQXNBSwsoZ4IWh/jeCy8jqE6qBI17t9vC3PFnaPw+4IeuUFpvOfmDNXuSI2UxX1sQVCc3giaePNkN5nYZgurktgetS/ukXJ0gqE56CWoOBNXJlaDMZhqSGa6gaTcQVCc07QaC6oSm3UBQndC0GwiqE5p2A0F1QtNuIKhOaNoNBNUJTbuBoDqhaTcQVCc07QaC6oSm3UDQ9rS4S07TbiBoe1zht/dl/HEQtD37jOLPgqbT279ulk0QtD0u+NVM0L/b0gjaniaChmeOvP/TDY2g7XHpc5Ivy3AI+g8Ebc/xvOOnId7FgnIMOjyzfm4WRLhaKuGX/30J20O9x5b45VtbAEFl3oYXCfqtbld+J+aR/qVYBkFlPgra8DJTpaC3HxgEQWW+COo+jsr5Zaaa7lAQtENfOnDHI6jM1yF+9J0kd/z8OemSL17XcE6JoCowKWi8fN2loF9AUBW8FrTBWcpbQbcl57w/Cbp3qq5yheViyqyUcLkhXFPLS/kMgsrMDO/jEP/75dSDPllhOU3p0pTRAXa+CmPdCd3DoGZk1o8tQfd+K7Jy78VDD/pkheXcYh8pH3eWuZmNL3AhqIwxQZMeNLmIEBt27kGT9/GFA38naPSSedoOBJUxLOipMz0sSnra49fTSslpyrOSh5nRMWj79ZYRVMauoHFHt79GZzJ+H5xDwsJKyWnK9AwoehfVVSrlMwgqY1HQTafIqrflPtyCXo2FoDIWBb0rBEEXYkVBzYGgMuYEZTZTy8z6sSbokiCoDIIqAEFlFAlaMXovujMQVEaToDffFz9ZAgSVQVAFIKiMOkG3n+78/f7JcaOSs/ivmfWjTdDw14bdvaClQkyCoDLKBC3ORY6SnDrWJXYPgsooE9TfCJp3rGvsHQSVmSpo6apSLGjpwpND0IaZ9aMovKoe1MfzjTVt/RcQVEZReJWCegRtl1k/isKLLjNVnsUr2vhPIKiMovB+4/f+kK9H0AGZ9aMoPO4kjc+sH0XhIej4zPpRFB6zmcZn1o+t8GxtbTUIKmMrPFtbWw2CytgKj2eSGmfWj63weKqzcWb92ApPFNR2X4qgMrbCSxcPO26LeiGQ2+jqw89Ttmw4BJWxFd6+ftLebfobQSsKbJ+y51a0zqwfW+Ht85ld/Gb/yvvnKyyHhFG+89LKR7a06FZriCGojK3waob4JyssnyeoHC9ZhrzKtI6vYc3KrB9b4cWzmreXYyLz/iZzM5cwiTg6xUoljD48VRmV16b5EFTGWniXPeje6Z170OR9FHNSXCbokcjl30XvGeL7Yi28W0EfrbCcjuYFQdPvXCYoQ3x/rIV3fQzqHq6w7LPzobiO9CQppEyL5iSpN9bCC4edyQHnOY0pEFTGWnin7UXQVluhE2vh1QhqDgSVsRYes5maZtbP4uHZAEFlFg/PBggqs3h4NkBQmcXDswGCyiweng0QVGbx8GxgQ9A5V04QVAFGBB1f5eC6QABBa2qFaVgSNJ6gM6zWeOrjErdmjGFIUGHZwd61lt/BIOwI6sb3oEmFCDoFO4LOGOJdXiGCjgZBL2vNJv3i53C+CTqMsK3jqnTHetsh3C+NBa+w0eY6zuIHVg07Nlpdg6A2Wmo5bDT78TjYREGPYw0Yh40W507Sn8XGTkDQWvITS/PYCGNOs9tom5TlzukWCaMLFtsGQf8QFttmP53Mzu6eFqEHXVujC4tt48LFB3d8IodSupFbl/ZtkqdY3AmjsNg2gqAXyU/vdIWta2t0YbFtZEH3gd9F69Lto3/yLkuSrRm+1ZFckXZxku27cspXAYGArbaJ5w+EBe6O488gjA+BhQNVl77bLM0zJILG02hc/sMl2b8ZamsnjMVW2wTXyj3oLmv8VhB0y59muBD0mEWDoEOx1TZ3gubJ5B60mP9K0KAogg7FVttcCpov5f07EtilS0fzYgZ/COqqh/hTytdhQQlbbRObV7oOGk6B9g/3tMd3xb+alB0THKdA8RAfSjm+K6Z8GxaUsNU20SCb/WaaRcLogq22QdA/h6222QRlNtPfwVbb2NraahYNqwm22sbW1lazaFhNmN82Twbs+VvbhUXDasL8tnlyyjN/a7uwaFhNmN82CLpqWE2Y3zbZ1fOKtOHdKifyK8TQi/ltc9yGeSxopy0azjKBdGB+2xxdZ4WgyU3F+dveiGUC6cD8tnkkaDazY40RXsFO0MuAtnE3BNPuEu5pj4LHBNCdFWLoxQhBb7+O57ffFpV1mivs3BVi6IUOQX+vT0+Sako3wQox9MKyoAVdbbJCDL1QIWg+/fi6qNMkYvMsEUQndAhalbAqhUkWDasJCKqARcNqggJBmc20aFhNmC/opKI0sWhYTUBQBSwaVhMsCsozSX8Ik4L2KHomi4TRBQRVwCJhdMGooNsAvxf9drhXYoaSzVCJUUG31+Su54s7n4/64n5NhaAyhgVNp4a+qkeJGUo2QyV2BU2c3N/EiykX11tO79+nSysX85WTtAVBZYwJGl1ccsd7lyzdnUyOcvtH2Yd7eff5Skkag6AytgSNCsxM87lF4UQq7nadT+c7pyW59ENXSNLnz5UjqMzKgib1Fx17LmgXRRFUxqqg6fl8QdDSesvPh/iiwwzxA1lDUBed3rj8fCga8Y+XvaTbfMehQZykfUBQxqqgnYqewyJhdAFBFbBIGF0wKSizmf4OFgVdDlpIBkEVQAvJPGibtyMrzX8HLSTzRNA3mR6n/ovQQjIIqgBaSOahoLVLgJyqiG7erHP23QqaQ+aZoPHNwGdVuPNHsENzyIwQNLrf+LjSPwHtIfPiGPR5D5odFbA/MmgQGfeALcezTC5bFtmvsm53Q2iPJvzG99N4XZVvwWWRG0JzNKEwDbM6HydJV9AcTTiOJD9dZnpewPLQHE0IB5/felAug56gPZrAnaRe0EJNQNBe0EJNYDZTL2ihqdD8d9BCU6H576CFpsIzSXcsEoZVeKrzjkXCsAqC3rFIGFbpIWh2xyDMSbGJ1e1ehC6CpkWf5qk2rbOisG/1IehUZghqC6vbvQh9BE3G9CCoi1ZDLi+tHCaoXiy7XCis6+rLCDqVhs1/zIAOBibzosOMwDAHNV9aef+29F28vcLSyln2UpI3Yb3JBK1oKWhSoMt/S38UzAxnV3FJu9lJJXmJzqcZknf7j7eRIuhUhgia9WGyoOeScrleC/paUQSdSg9Bs7Mil3zqRUHrll32/5t2VVhRUIZ4o3TpQdMLn7k+x0VS54snSfGIf7z48OGpsK6rLyPoVLoN8cvs2UXCsAqC3rFIGFZpLyizmaAhzQVdjkXDsgKC3rFoWFa4af4nA/aie3LRsKxwJ2htwqoUJlk0LCsg6B2LhmWFe0GrFyw536xZYt8uEYRdKnvQV4KucaFpiSDsUieoqxrio7uUv0xL7NslgrBLlaDuPqHfhEwOWpfYt0sEYRd3w5bmPqHLFmuuk9oASwSxLE+PQR2CwkjenyQtczt+hRjWxQXvuMwECjlPP75Oi6AwFO4krRrWIiDoqmEtArOZVg3rD7Lonlw0rD/Ionty0bD+IDyTBKrhqU5QDYKCahJBo0v7rvZKf4vq7ZQLo4kFlZa+iZL0qb59DQi6CpGgydTQ6pv5Laq3Uy6MJpncXBI0TC/9vX+1wrKYcn98KtQQzyFweSmPwwL7RNOWgxLxZOYwQzRMHP2wwvL2cqTc9TtqKBb9Iqw3mUAtkaDhk8OOK328jw2P+76k/LicXOzQhfpQfbmUZwHBQoSuMx7ik87zQlAfv0YZTooWBS3UIJfyJCBYiMOLWDdXIWj1CsvFfC4e6bMaHEM87JQFPY3D2blL9rJlrDhJCqWkL+myy1FFrwOCRXCF30yzSBiwgaCgGmYzAYwEQUE1CAqqQVBQDYKCahAUVIOgoBoEBdUgKKgGQUE1CAqqQVBQDYKCahAUVIOgoBoEBdUgKKgGQUE1CAqqQVBQDYKCahAUVIOgoBoEBdUgKKgGQUE1CAqqQVBQDYKCahAUVIOgoBoEBdUgKKgGQUE1CAqqQVBQDYKCahAUVIOgoBoEBdUgKKgGQUE1CAqqQVBQDYKCahAUVPMfDjDQU6Gh7ZgAAAAASUVORK5CYII=" /><!-- -->
Fig. 3. The labelled training species tree for this example.</p>
<hr />
<p>Hence, <span class="math inline">\(a=0.25\)</span> and <span class="math inline">\(\psi=2\)</span> for edges 15–21, <span class="math inline">\(a=0.8\)</span> and <span class="math inline">\(\psi=0.5\)</span> for edges 10–13, and <span class="math inline">\(a=0\)</span> and <span class="math inline">\(\psi=1\)</span> for edges 1–9 and 14:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">attr</span>(perissodactyla.pgraph,<span class="st">&quot;ev&quot;</span>)[<span class="dv">1</span><span class="dt">L</span>]) <span class="ot">-&gt;</span> steepness</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">rep</span>(<span class="dv">1</span>,<span class="fu">attr</span>(perissodactyla.pgraph,<span class="st">&quot;ev&quot;</span>)[<span class="dv">1</span><span class="dt">L</span>]) <span class="ot">-&gt;</span> evol_rate</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>steepness[<span class="dv">15</span><span class="dt">L</span><span class="sc">:</span><span class="dv">21</span>] <span class="ot">&lt;-</span> <span class="fl">0.25</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>evol_rate[<span class="dv">15</span><span class="dt">L</span><span class="sc">:</span><span class="dv">21</span>] <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>steepness[<span class="dv">9</span><span class="dt">L</span><span class="sc">:</span><span class="dv">13</span>] <span class="ot">&lt;-</span> <span class="fl">0.8</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>evol_rate[<span class="dv">9</span><span class="dt">L</span><span class="sc">:</span><span class="dv">13</span>] <span class="ot">&lt;-</span> <span class="fl">0.5</span></span></code></pre></div>
<p>The <strong>PEM</strong> is obtained as follows:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">PEM.build</span>(</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>  perissodactyla.pgraph,</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>  <span class="at">d=</span><span class="st">&quot;distance&quot;</span>,</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>  <span class="at">sp=</span><span class="st">&quot;species&quot;</span>,</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>  <span class="at">a=</span>steepness,</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>  <span class="at">psi=</span>evol_rate</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.PEM</span></code></pre></div>
<p>In addition to the phylogenetic graph, function
<code>PEM.build()</code> needs arguments <code>d</code>, the name of the
edge property where the phylogenetic distances are stored,
<code>sp</code>, the name of the vertex property specifying what vertex
is a species, as well as the user-specified steepness and evolution
rate. When the vectors given to arguments <code>a</code> or
<code>psi</code>, have smaller sizes then the number of edges, the
values are recycled. The default values for <code>d</code> and
<code>sp</code> are those produced by
<code>Phylo2DirectedGraph()</code>, and can therefore be omitted in most
cases. The object that <strong>MPSEM</strong> uses to store
<strong>PEM</strong> information is rather complex and we will hereby
not browse through it. Method <code>as.data.frame</code> can be used to
extract the eigenvector from the <code>PEM-class</code> object. For a
set of <span class="math inline">\(n\)</span> species, that method
returns a matrix encompassing at most <span class="math inline">\(n-1\)</span> column vectors that can be used in
models to represent phylogenetic structure in traits. Here the
phylogenetic patterns of variation described by two eigenvectors of the
<strong>PEM</strong> we calculated above:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>),<span class="dv">1</span><span class="dt">L</span>,<span class="dv">7</span><span class="dt">L</span>))</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>,<span class="fl">2.1</span>,<span class="fl">4.1</span>,<span class="fl">2.1</span>))</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="do">## Singular vectors are extracted using the as.data.frame method:</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="fu">as.data.frame</span>(perissodactyla.PEM) <span class="ot">-&gt;</span> perissodactyla.U</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="fu">plot</span>(perissodactyla.tree.train, <span class="at">x.lim=</span><span class="dv">60</span>, <span class="at">cex=</span><span class="fl">1.5</span>)</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">y =</span> <span class="dv">1</span><span class="dt">L</span><span class="sc">:</span><span class="fu">nrow</span>(perissodactyla.train), <span class="at">ylab=</span><span class="st">&quot;&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;Loading&quot;</span>,</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>     <span class="at">x =</span> perissodactyla.U[,<span class="dv">1</span><span class="dt">L</span>], <span class="at">xlim=</span><span class="fl">0.5</span><span class="sc">*</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>     <span class="at">axes=</span><span class="cn">FALSE</span>, <span class="at">main =</span> <span class="fu">expression</span>(<span class="fu">bold</span>(v)[<span class="dv">1</span>]), <span class="at">cex=</span><span class="fl">1.5</span>)</span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>)</span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">0</span>)</span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">y =</span> <span class="dv">1</span><span class="dt">L</span><span class="sc">:</span><span class="fu">nrow</span>(perissodactyla.train), <span class="at">ylab=</span><span class="st">&quot;&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;Loading&quot;</span>,</span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>     <span class="at">x =</span> perissodactyla.U[,<span class="dv">5</span><span class="dt">L</span>], <span class="at">xlim=</span><span class="fl">0.5</span><span class="sc">*</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>     <span class="at">axes=</span><span class="cn">FALSE</span>, <span class="at">main =</span> <span class="fu">expression</span>(<span class="fu">bold</span>(v)[<span class="dv">5</span>]), <span class="at">cex=</span><span class="fl">1.5</span>)</span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>)</span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAtAAAAGACAMAAACz76XEAAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAARSklEQVR4nO2dgXKkIBAF8f9/+uqyqwKigggLb7qrLmYjMDJ2DOsJ6xYAIdyvDwDgTRAapEBokAKhQQqEBikQGqRAaJACoUEKhAYpEBqkQGiQAqFBCoQGKRAapEBokAKhQQqEBikQGqRAaJACoUEKhB4aodPj/mgfpnkEKMet/yROz6c3nboikTE5vkJ3s6AtCG2e9QqtcXpWoXuMODQypsd6QdM4PX+9cdtvaeNYMCCCQn+/ax8LRsQFm9lxh2/ah4IRETo9nd4TCGVMjz53bnvRpzM6+QJYEBrEQGiQAqFBCoQGKRAapEBokAKhQQqEBikQGqRAaJACoUEKhAYpEBqkQGiQAqFBCoQGKRAapEBokAKhQQqEBikQely0zg1r25lH69wgtHm0zg1Cm0fr3CC0ebTOjZjQ+R9IoHUaa9DKhJrQ1SW1Tu8fN7/kWj1e11NtvMTdPELr4W4+RkUrE98Fr1t/dExvof//erp1adV1hdXgl3b96AK3br7Fvl+d24okq8+Dc/4mVaDbofTAO4stje45hl62j3YKhQ5XDnYu+KHzC6wV3Hn1ebhd1X7CPl0QnKbxhXZ3eOESV+j4iNz2LjIp9EX1eUDoZmE6tOMJfbhCL4GT/r4lS+hJlUboZmE6tHMcQzsX6e2XvBly7JtD9XlgDN0sTId2onHHx2tvc7hCB6PuZb3bsw5D4lamPPfc5WgWpm87WieqAu5DNwvTtR2tE9UOrTx16s0vhIY8tHKK0ObRyilCm0crpwhtHq2cIrR5tHKK0ObRyilCm0crpwhtHq2cIrR5tHKK0ObRyilCm0crp2JC58/69qs8Px4FtLqvJvSjFl302hZaPUboZwcyEzw+2ixM83aOM1b8WSlBH70n+/8KrfO9/X0Sk795wL9ZmPbt5M/6/ja36rm5HXzzLXLTytgwBatZmA7tOG9zM217m+4dNuyPqT9FLlsZHxdtTwtoEFxwJhA6fxmDvFnfV0KfXeDnUhqhm4Xp0M5xDH016/vyCh3XS7YyAQjdLEyHdvJnfW8/DBoOZn2Hby1nnfzNGLpZmM7taJ2o53CXo1mYvu1onagKuA/dLMxNoQwaHqNVtHLaqTd5Qr/UDhShlVOENo9WThHaPFo5RWjzaOUUoc2jlVOENo9WThHaPFo5RWjzaOUUoc2jlVOENo9WThHaPFo5FRP66QMfWie1DK2+qwmdE+fFekPBL/OyPrje/Ck2hG7P02eAFfq+49o/C/0J806hXKHvlzFYByfRUgX++gVz8XiWxnQ9vcS1n63y1/xLhXLG0N9iGcsYuGXb5/aXbts5FS7aFlfUwD3OQ1mYlwrFRS5mfaeu0MfGvrW2Il6zOYc8EAj9x+xCp18nr9BLpLSLKuxCH6/lE4DQf2gKnbGMQeiuP+TYfh/mgjH0H5OPodNDkG3fxTIGyTeF3pBjvvPMXY7/TH6X46qK1onKgPvQy3Y3q3WnEHpctPLUqTe/EBry0MohQptHK4cIbR6tHCK0ebRyiNDm0cohQptHK4cIbR6tHCK0ebRyiNDm0cohQptHK4cIbR6tHM4m9M2K/vmzvkueAunzROKv0OrVZELfVclvouixJpfaqKDVHYTO2Te0ybVPS47Zq2d0mzvXW+iMWd/BR9KH070Pk78DoQcbeVQ/zz5Wd6r4nPUeHeomdPas7/Aj6ZfgVbRv6CFH/fh+qO5U8b1C9TC68xU6bXJqiqwv9KH69gdsZKGj7fMW5sd5/zqEeqfQeZVoGYO8D6+/FDoOg9Bjoya09/p+1rc3rHCh0KlfhCDaUAYg9I6i0Pmzvvc3fvsVentTuFXfGo03o8AYekdsDH0ynlaHuxw7Wnc5jArNfWgPqfvQSiemJ1p569QbhB4XrbwhtHm08obQ5tHKG0KbRytvCG0erbzNLvTFw/6Qh1beJhf6xepm0cobQptHK28IbR6tvCG0ebTyhtDm0cqbmNDc7ihHK11qQtc1kWxzdHjabmf6p+3i6konJxOeh96Z/3nouPraxP0yBluRYMUCF1QYbcWCFMxY2RGYsRJXL1jGIPi3OR+VHP9ku2j7vIX58c9ph1DvFLqoHs76Tl2ho2B757fPlV0Qel7EhPaaSF6hl1DpYMkN78LtturvHFJjEHpHV+icZQyC7T5qXkcu01yhGUNviI2h/5rIXsbAH24v3j7f+CluanOXY0frLkdZc1cX4Ekuzl+4D70jdR+6sLmrrs9wZX4NrZ526s0vhIY8tNKO0ObRSjtCm0cr7QhtHq20I7R5tNKO0ObRSjtCm0cr7QhtHq20I7R5tNKO0ObRSjtCm0cr7WpCv/d0itZ5vkCro2pCn7SS0XRcZK5H7irQ6qMRoX/YzvCcd/QHz4NXhxQVOvwQ+u1TNhMfQrhV84psM1Zc0Fiy3vxcPES7fulGfUhRoT/f7FNSti8nc8DXvXsR51W5rDc9Z52pn9xVfij1IUWF/vzlCl3+tr9faY/1DtJuPzyvNz2nQt/sb8ALIQWEjlbx98U8XqGD10HAXKHllEboVmGeCn18+R0zbGPoWOjUpO5I6OSQY57J4AUgdKsw7wntLwW2rAOQ7VJ7Ngc8WDRsLbKXTNYTgDF0qzDvCF3foDFOs8RdjsowCP0TLvLHfeiqMK8LDTlo5Q+hzaOVP4Q2j1b+ENo8WvlDaPNo5Q+hzaOVP4Q2j1b+ENo8WvlDaPNo5Q+hzaOVP4Q2j1b+ENo8WvmbTGiX4lDg2TFWHNbUaPVvNqHvf5ZsJb+bx5JaJ/zIOE/bvRHPiNA1mBW69/PQr8QTFbp0GYNkyaiVYOJKspU5OTv+3jNW3oknKvTnm/xlDMKSLizp9pLx1MKwlTk5Ffpmf6vjQOh16y8IU7iMQWz9stkaXtJjk4/vSucDoVuFqRV63xyurQmhA6WjK/S3+uJX8DaR15OD0K3CvCt00TIGYcmjwrvJ3hhaZFEDxtCtwrwndPEyBmFJf5ARDDm8GDqLGpweP3c5KsO8I3Rp1MKST4qPzXlnuA9dFwahf4JUZ7SFhhy08obQ5tHKG0KbRytvCG0erbwhtHm08obQ5tHKm4LQqQf9IRutvAkIXRQDDmjlDaHNo5U3hDaPVt4Q2jxaeVMTmneHxWilS03okyK3TT9/jml6H6bvQIARod+IXVxyEnh8tFWYF4UunvW9FfTqHad779XCpqd+zv/0wHnAvzLMy1foog+vd4tfzPmbqIIXxh2+mZGz42YKVm2Yl6/QRR9evyYznKwV/tCrH/zeLF2vY69zKvTN/lbHgdDrvl3G9aJ6vEIHr73WAlkjob0I8T7v9bRKI3SrMJVCe0W+g7GCD68PRhcJocN9LhJac8hxs7/VcSB0okjprO8lev8XRAveFG4lw6YF3xQyhq4N847Qj0tb5TRL3OWoDIPQP+E8S9yHrgvzutCQg1ZOEdo8WjlFaPNo5RShzaOVU4Q2j1ZOEdo8WjlFaPNo5RShzaOVU4Q2j1ZOhxI6g9YHahCtnI4kdMd2YEcrpwhtHq2cIrR5tHKqJnTpOFvrbD4i5734PG9fcg+zsku9hL6cOuISr/LKPi0yBff96P1gdA2ZR1nbpa5C5+28FtoQGX/1/M3g5B1kdZd6Cx2uyxFOk0qur5FeyOPzZf/Azb36mozvdK5kyUm4Fzq34AhkCl1U+nGYuna8UdEm2F5l1S7Yt5l7qBAIvXieuviLC6pPaDRCNwtT145v72FarDsT+vMlqnAh9P5nCqFHRFTooHhS4UjouP1zofcB2FpVXmjG0A/D1LXjOXiUdl+CJhxdJCt4jbnsIceh5CzcHzB3OZ6GqWrnI1hyfQ23OuftSy7dGI1R9rd8/pBja2Xflyw5CRkHzH3op2Fq2pkk48OhlbdOvUHocdHKG0KbRytvCG0erbwhtHm08obQ5tHKG0KbRytvkwvNfMN6tPI2u9ANYlhDK28IbR6tvCG0ebTyhtDm0cobQptHK29iQj+926F1UsvQ6ruY0PsDnGURXeK77IOYHB4fbRimtJ1zoevatcR933nA/2mY4nZOhfYneCfngIdL0oTTvZP10kUEuO0GU7Cehilv53rI8f1mfe38V+G4xGXUSxVR4F7o3IIjkCl0UenHYcrbCYReh0Xx2COc0r1pmvxlCOqtJRNFJrle5YDQzcKUt3NoNyl0UDjpZLnQOkojdLMw5e1kCJ2aA14+5Eg6P8UZvoUxdLMw5e0cXkdzsk/mgHs/XNu5rbcPVYKFxebnvhvc5XgapridOZI8Nhk55D700zCl7UyS5aHRymGn3iD0uGjlEKHNo5VDhDaPVg4R2jxaOURo82jlcDahY15q1zJaOZxM6F7tWkIrhwhtHq0cIrR5tHKI0ObRyiFCm0crhwhtHq0cigkdf9wg3MPTdg3D1LbrP5r/fuCcU/9mvD7wPHS7MNXtHoSGW5ix0ixMfbub0P6SA+n1C3LWNkg0prfEwb3QuQVHIFPootKPw9S16/13eDRF8Gv1ktgsqX1+4yfrF0TVU0VmAaGbhalrN1BzOXXwWyV/bYNEY16F4NX6ZYpTv4LQzcLUtfvVLbxGngsdtJCU8bHQcynNGLpZmLp23f7vVui8tQ2Wvy5fNZYUWm3IwV2Ox2Gq2k3p9jes8F8th823cuJN4fcj6cPG9JY4yDhU7kM/DVPT7iQZHw6tvHXqDUKPi1beENo8WnlDaPNo5Q2hzaOVN4Q2j1beENo8WnmbXWiWNKhGK2+TC907hiJaeUNo82jlDaHNo5U3hDaPVt4Q2jxaeRMTutOsbykHpDojJ/TnGxf8sE2othE6wuOjDcPUkBS6XSgZeMC/XZgqAqHfmPV9WtKbLRDM8/af8+/Q4Ze4PVSmYD0NU4c/hvbk3q1eEpsltc8/6H3fXnLNwx4h2fQk3AudW3AEMoUuKv04TB3hfKtr3db/NF+PbL+2hi0u6erOG4NtEc5aGR2EbhamHhcovFwJHRyZV+GgdFLoRITzVsYGoZuFqedr163Q2bO+k/WcP/KIIjjBIQdj6Idh6kmOC/xXy2HzrZjxpnBrJdyEU8G9QJNwf6jc5XgaBn5AxrnhPvTTMNAfrXPTqTdaSdNC69wgtHm0zg1Cm0fr3CC0ebTODUKr4LpjuTcI3ZzLFF/tbLGvmseBO/UGoZuD0FUVEXo0ELqqIkKPBkJXVUTo0UDoqooIPRoIXVURocEyCA1SIDRIgdAgBUKDFAgNUiA0SIHQLYkfFcvKdmoJkvTL6Fm0R9EKeL03V5152huEboiL8pv1KGRYKWoieOm2L8+jFfB6b64687g3CN2O4znKyHZY6ejs6c5H0Qp4vTcFPS3oDUK34+YKmlPp5jRfF20qdH1vrjrzvDcI3YLPTIvj+Xtf6Kur91vntlFvrjrzvDcI3Y72Qrs3FMikudDhW8Qn0aIG4WWik5I3Qa5E6OPlrTRaAa/35qozz3uD0A3ZRn5u/0FhpWjwGOx0VxUzoxXwem+uOvO4Nwjdkm0tzW+as7IdVkrfh3brQpThEpYPohXwem+uOvO0Nwjdg0dZvqx0tbP1OX29N497Wl8cYGwQGqRAaJACoUEKhAYpEBqkQGiQAqFBCoQGKRAapEBokAKhQQqEBikQGqRAaJACoX/O+SmY8eT8ujcz5kyMGwUmO0O/7s1k6VLk19e0d/l1b2bMmRjbLNDP9Ll985k8+plBGv5wYH7dm8HTYwF/Euj2/Xb2Aw+W8U/Yr3szen4MECkQr1Pk9m8mFLp7b0bPjwFCBcLNEiiwhBP9h+TXvRk+QfocFXCHP9L7vtH5dW9myJE46xIr4funfeP9yX57ea8G/Lo3o+cHPGa8L31Om97IpMcEE1yhC+AKDXAHQoMUCA1SIDRIgdAgBUKDFAgNUiA0SIHQIAVCgxQIDVIgNEiB0CAFQoMUCA1S/AMTeSF+jM8p7QAAAABJRU5ErkJggg==" /><!-- -->
Fig. 4. Example of two eigenvectors obtained from the training species
phylogeny.</p>
<hr />
<p>The pattern shown by the first eigenvector essentially contrasts
Equids and the other odd-toed ungulate species whereas the pattern shown
by the second eigenvector essentially contrasts tapirs and
Rhinocerotids.</p>
</div>
<div id="estimate-weighting-parameters-empirically" class="section level2">
<h2>Estimate weighting parameters empirically</h2>
<p>Because users do often not have information about the best set of
weighting function parameters to use for modelling,
<strong>MPSEM</strong> as has a function called
<code>PEM.fitSimple()</code> that allows them to empirically estimate a
single value of parameter <span class="math inline">\(a\)</span> for the
whole phylogeny using restricted maximum likelihood. Function
<code>PEM.fitSimple()</code> does not estimate parameter <span class="math inline">\(\psi\)</span> because the latter has no effect
when its value is assumed to be constant throughout the phylogeny. A
function to estimate different sets of weighting function parameters for
different portions of the phylogeny has yet to be developed. That
function requires a response variable that will be used to optimize the
steepness parameter (here the <span class="math inline">\(\log_{10}\)</span> neonate weight) as well as
lower and upper bounds for the admissible parameter values and is called
as follows:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">PEM.fitSimple</span>(</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>  <span class="at">y =</span> perissodactyla.train[,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>  <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>  <span class="at">w =</span> perissodactyla.pgraph,</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>  <span class="at">d =</span> <span class="st">&quot;distance&quot;</span>,</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>  <span class="at">sp=</span><span class="st">&quot;species&quot;</span>,</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a>  <span class="at">lower =</span> <span class="dv">0</span>,</span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>  <span class="at">upper =</span> <span class="dv">1</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.PEM_opt1</span></code></pre></div>
<p>If other traits are to be used in the model (here the <span class="math inline">\(\log_{10}\)</span> female weight), they are passed
to argument <code>x</code> as follows:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">PEM.fitSimple</span>(</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>  <span class="at">y =</span> perissodactyla.train[,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="at">x =</span> perissodactyla.train[,<span class="st">&quot;log.female.wt&quot;</span>],</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  <span class="at">w =</span> perissodactyla.pgraph,</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>  <span class="at">d =</span> <span class="st">&quot;distance&quot;</span>,</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>  <span class="at">sp=</span><span class="st">&quot;species&quot;</span>,</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>  <span class="at">lower =</span> <span class="dv">0</span>,</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>  <span class="at">upper =</span> <span class="dv">1</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.PEM_opt2</span></code></pre></div>
<p>The results of the latter calls are a <strong>PEM</strong> similar to
that obtained using <code>PEM.build()</code>, with additional
information resulting from the optimization process. It is noteworthy
that estimates of the steepness parameter (stored as element
<code>$optim\$par</code> of the <strong>PEM</strong> objects) and,
consequently, the resulting phylogenetic eigenvectors, will be different
depending on the use of auxiliary traits. In the example above, for
instance, <span class="math inline">\(a\)</span> was estimated to 0 by
<code>PEM.fitSimple()</code> when no auxiliary trait is involved (first
call) and to 0.08 when the female weight is used as an auxiliary trait
(second call).</p>
</div>
<div id="phylogenetic-modelling" class="section level2">
<h2>Phylogenetic modelling</h2>
<p>To model trait values, <strong>PEM</strong> are used as descriptors
in other modelling method. Any suitable method can be used. For
instance, package <strong>MPSEM</strong> contains a utility function
called <code>lmforwardsequentialAICc()</code> that does step-wise
variable addition in multiple regression analysis on the basis of the
corrected Akaike Information Criterion (AICc) <span class="citation">(Hurvich and Tsai 1993)</span>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">lmforwardsequentialAICc</span>(</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>  <span class="at">y =</span> perissodactyla.train[,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>  <span class="at">object =</span> perissodactyla.PEM_opt1</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>) <span class="ot">-&gt;</span> lm1</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="fu">summary</span>(lm1)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = as.formula(paste(p1, p2, sep = &quot;&quot;)), data = df1)
## 
## Residuals:
##       Min        1Q    Median        3Q       Max 
## -0.122057 -0.053781 -0.004971  0.055333  0.186364 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  4.39975    0.02740  160.58  &lt; 2e-16 ***
## V_2          1.31105    0.09491   13.81  7.7e-08 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.09491 on 10 degrees of freedom
## Multiple R-squared:  0.9502, Adjusted R-squared:  0.9452 
## F-statistic: 190.8 on 1 and 10 DF,  p-value: 7.699e-08</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">lmforwardsequentialAICc</span>(</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>  <span class="at">y =</span> perissodactyla.train[,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  <span class="at">x =</span> perissodactyla.train[,<span class="st">&quot;log.female.wt&quot;</span>,<span class="at">drop=</span><span class="cn">FALSE</span>],</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>  <span class="at">object =</span> perissodactyla.PEM_opt2</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>) <span class="ot">-&gt;</span> lm2</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="fu">summary</span>(lm2) </span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = as.formula(paste(p1, p2, sep = &quot;&quot;)), data = df1)
## 
## Residuals:
##       Min        1Q    Median        3Q       Max 
## -0.024813 -0.010155 -0.003962  0.008556  0.040311 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)   -2.66734    0.12638 -21.106 7.37e-07 ***
## log.female.wt  1.23818    0.02211  56.001 2.18e-09 ***
## V_1           -0.91653    0.02742 -33.432 4.77e-08 ***
## V_9           -0.09882    0.02362  -4.183  0.00579 ** 
## V_8            0.08489    0.02378   3.570  0.01178 *  
## V_3            0.07163    0.02378   3.012  0.02364 *  
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.0236 on 6 degrees of freedom
## Multiple R-squared:  0.9982, Adjusted R-squared:  0.9966 
## F-statistic: 648.5 on 5 and 6 DF,  p-value: 4.126e-08</code></pre>
<blockquote>
<p>It is noteworthy that in order to pass a single auxiliary trait to
<code>lmforwardsequentialAICc()</code>, it is necessary, as exemplified
above, to set <code>drop=FALSE</code> to the extraction operator
(<code>[]</code>) to keep the <code>data.frame</code> property of the
object. Hence, it is crucial that the auxiliary trait values be
referenced with the same variable names in all <code>data.frame</code>
objects involved in the analysis.</p>
<p>Since, the extraction operator drops the <code>data.frame</code>
property on its output object by default whenever a single column is
selected, it is mandatory to set <code>drop=FALSE</code> when a single
variable is (or could be) selected. A <code>data.frame</code> object is
required for making predictions using the resulting linear model.</p>
</blockquote>
<p>To obtain predictions, we need to calculate the locations of the
target species with respect to the phylogeny of the model species. This
is accomplished by <code>getGraphLocations()</code>, to which we give
the tree for all species (model + targets) and the names (or indices) of
the target species. Then, we use the <code>predict()</code> method for
<strong>PEM</strong> objects. The latter takes, in addition to the
<strong>PEM</strong> object, the locations of the target species as
obtained by <code>getGraphLocations()</code>, an <code>lm</code> (or
<code>glm</code>) object involving the eigenvectors of the
<strong>PEM</strong>, and a table of auxiliary trait values for the
target species, which can be omitted if no auxiliary trait is present in
the linear model.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">getGraphLocations</span>(</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>  perissodactyla.tree,</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>  <span class="at">targets =</span> <span class="st">&quot;Dicerorhinus sumatrensis&quot;</span></span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.loc</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a><span class="fu">predict</span>(</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>  <span class="at">object =</span> perissodactyla.PEM_opt2,</span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>  <span class="at">targets =</span> perissodactyla.loc,</span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a>  <span class="at">lmobject =</span> lm2,</span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a>  <span class="at">newdata =</span> perissodactyla.test,</span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a>  <span class="at">interval =</span> <span class="st">&quot;prediction&quot;</span>,</span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a>  <span class="at">level =</span> <span class="fl">0.95</span>) <span class="ot">-&gt;</span> pred</span></code></pre></div>
<p>Here, the predicted neonatal weight for the Sumatran rhinoceros is
26.7 <span class="math inline">\(\,\mathrm{kg}\)</span> and the bounds
of the <span class="math inline">\(95\%\)</span> prediction interval are
17.6 and 40.4 <span class="math inline">\(\,\mathrm{kg}\)</span>, while
the observed value was actually 35 <span class="math inline">\(\,\mathrm{kg}\)</span>.</p>
</div>
</div>
<div id="cross-validating-pem-predictions" class="section level1">
<h1>Cross-validating <strong>PEM</strong> predictions</h1>
<p>Here is and example of how to perform a leave-one-out
cross-validation of a data set using the <strong>R</strong> code from
the previous two sections. Predictions will be added to table
<code>perissodactyla.data</code>:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>  perissodactyla.data,</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>  <span class="at">predictions =</span> <span class="cn">NA</span>,</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="at">lower =</span> <span class="cn">NA</span>,</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>  <span class="at">upper =</span> <span class="cn">NA</span></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>) <span class="ot">-&gt;</span> perissodactyla.data</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>jackinfo <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="dt">L</span><span class="sc">:</span><span class="fu">nrow</span>(perissodactyla.data)) {</span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>  jackinfo[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a>  <span class="fu">getGraphLocations</span>(</span>
<span id="cb25-12"><a href="#cb25-12" tabindex="-1"></a>    perissodactyla.tree,</span>
<span id="cb25-13"><a href="#cb25-13" tabindex="-1"></a>    <span class="at">targets =</span> <span class="fu">rownames</span>(perissodactyla.data)[i]</span>
<span id="cb25-14"><a href="#cb25-14" tabindex="-1"></a>  ) <span class="ot">-&gt;</span> jackinfo[[i]][[<span class="st">&quot;loc&quot;</span>]]</span>
<span id="cb25-15"><a href="#cb25-15" tabindex="-1"></a>  <span class="fu">PEM.fitSimple</span>(</span>
<span id="cb25-16"><a href="#cb25-16" tabindex="-1"></a>    <span class="at">y =</span> perissodactyla.data[<span class="sc">-</span>i,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb25-17"><a href="#cb25-17" tabindex="-1"></a>    <span class="at">x =</span> perissodactyla.data[<span class="sc">-</span>i,<span class="st">&quot;log.female.wt&quot;</span>],</span>
<span id="cb25-18"><a href="#cb25-18" tabindex="-1"></a>    <span class="at">w =</span> jackinfo[[i]][[<span class="st">&quot;loc&quot;</span>]]<span class="sc">$</span>x</span>
<span id="cb25-19"><a href="#cb25-19" tabindex="-1"></a>  ) <span class="ot">-&gt;</span> jackinfo[[i]][[<span class="st">&quot;PEM&quot;</span>]]</span>
<span id="cb25-20"><a href="#cb25-20" tabindex="-1"></a>  <span class="fu">lmforwardsequentialAICc</span>(</span>
<span id="cb25-21"><a href="#cb25-21" tabindex="-1"></a>    <span class="at">y =</span> perissodactyla.data[<span class="sc">-</span>i,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb25-22"><a href="#cb25-22" tabindex="-1"></a>    <span class="at">x =</span> perissodactyla.data[<span class="sc">-</span>i,<span class="st">&quot;log.female.wt&quot;</span>,<span class="at">drop=</span><span class="cn">FALSE</span>],</span>
<span id="cb25-23"><a href="#cb25-23" tabindex="-1"></a>    <span class="at">object =</span> jackinfo[[i]][[<span class="st">&quot;PEM&quot;</span>]]</span>
<span id="cb25-24"><a href="#cb25-24" tabindex="-1"></a>  ) <span class="ot">-&gt;</span> jackinfo[[i]][[<span class="st">&quot;lm&quot;</span>]]</span>
<span id="cb25-25"><a href="#cb25-25" tabindex="-1"></a>  <span class="fu">predict</span>(</span>
<span id="cb25-26"><a href="#cb25-26" tabindex="-1"></a>    <span class="at">object =</span> jackinfo[[i]][[<span class="st">&quot;PEM&quot;</span>]],</span>
<span id="cb25-27"><a href="#cb25-27" tabindex="-1"></a>    <span class="at">targets =</span> jackinfo[[i]][[<span class="st">&quot;loc&quot;</span>]],</span>
<span id="cb25-28"><a href="#cb25-28" tabindex="-1"></a>    <span class="at">lmobject =</span> jackinfo[[i]][[<span class="st">&quot;lm&quot;</span>]],</span>
<span id="cb25-29"><a href="#cb25-29" tabindex="-1"></a>    <span class="at">newdata =</span> perissodactyla.data[i,<span class="st">&quot;log.female.wt&quot;</span>,<span class="at">drop=</span><span class="cn">FALSE</span>],</span>
<span id="cb25-30"><a href="#cb25-30" tabindex="-1"></a>    <span class="at">interval =</span> <span class="st">&quot;prediction&quot;</span>,</span>
<span id="cb25-31"><a href="#cb25-31" tabindex="-1"></a>    <span class="at">level =</span> <span class="fl">0.95</span></span>
<span id="cb25-32"><a href="#cb25-32" tabindex="-1"></a>  ) <span class="ot">-&gt;</span> predictions</span>
<span id="cb25-33"><a href="#cb25-33" tabindex="-1"></a>  <span class="fu">unlist</span>(predictions) <span class="ot">-&gt;</span> perissodactyla.data[i, <span class="fu">c</span>(<span class="st">&quot;predictions&quot;</span>, <span class="st">&quot;lower&quot;</span>, <span class="st">&quot;upper&quot;</span>)]</span>
<span id="cb25-34"><a href="#cb25-34" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Warning in PEM.fitSimple(y = perissodactyla.data[-i, &quot;log.neonatal.wt&quot;], : No
## optimum found... Message from optim() - ERROR: ABNORMAL_TERMINATION_IN_LNSRCH.
## Status = 52</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">rm</span>(i, predictions)</span></code></pre></div>
<p>Because the result of <code>getGraphLocations()</code> includes the
phylogenetic graph without the target species <em>Dicerorhinus
sumatrensis</em>, which was removed from the tree given as the argument
<code>tpall</code> and can be found as its element <code>$x</code>, it
is not necessary to re-calculate the tree with the target species
dropped, as well as the phylogenetic graph, as we did previously for
explanatory purposes. Also, the internal information about each
cross-validation steps is stored into a list (hereby called
<code>jackinfo</code>), in order for the details of the analyses to be
accessible later on.</p>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAKgCAMAAABz4j/3AAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAX3UlEQVR4nO2djXajOgwG6fu/9D23u802wYBly+KTmDln2ySAbCuzBP+Ebl8Awmx3VwDgDAQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpFkv6Aawp1ufSfF+P/cuAcoSIujuaTsWgsInW78VXoJuh8EQFD44PJm19x0u5fcnPIJCL9tXjKBv5ewEtV8OwzM4Ppcd7j1Z2HmpCAq/2X797N59srQvBIVOtrdf3fuPlvT2g148XLF9/O4+YKSsv9eXr1/uJUA1tt2D7iM8Sl1WAtRgazzqPmQVCAp/2ZoPu49ZBILCH7aDx90HrQFB4Zvt8En3UUtAUPifxtS4/bAVICh8tdYWDR23AASFvQUICkrsJEBQEGLvAIKCDg0FEBQW07+esrUDgkIEPW/u3Bo3BIUJPgY3W2fVyTVuCAoTXF9dzq5xQ1CY4FLQo3cfQcHEwC0Vvg+7eOkwGIKCHftbdSHocUAEBTvegp7EQ1Cw4yzoWTgEBTu+gp5GQ1Cw4yqo0zfVEBReeAp6EQtBwY6joFehEBTs+Ak6tIBkbsdhEDQNboJeB0JQsOMlaEccBAUTnlOdw2vwpnYcBkHT4HMG7YqCoGDHfS7eoSgEhRcegnZeIiAo2HEQtPfvdyAo2JkXtK8LbykKQeHFtKB9Y6CmohAUXswK2jeJZCsKQeHFpKB9s/DGohAUXswJ2reMyVoUgsKLKUH71oGai0JQeDEjaN9CentRCAovJgRtmOpTFILCi3FBGx/1TkUhKLwYFrTRV/IqCkHhxaigjcEmt6IQFF4MCtoYrfcrCkHhxZigjelOx6IQFF4MCdpYL+JZFILCixFBGwvuXItCUHgxIGhjxbJvUQj6CPq+E+exoh5BYZTLd8L8VrV0R1AYxF3Q5vc7EBQG8Ra00YXvDoKgsMNZ0MYYaH8QBIUdvoI2JpEMQRAUdrgK2piFtwRBUNjhKWhjGZMpCILCDkdBG+tAbUEQFHb4CbrtHliDICjscBN0azwyBkFQ2OEl6NZ8aAuCoLDDSdDt4PFwEJ8dh0FQFXwE3Q6fGIIgKOxxEXQ7edYdBEGhgYeg2+nTziDde1l2HAZBVXAQ9OP82VpfiqAwyLygn37+v+Cuw9mRosw7DoOgKkwLuvPze8Xd9v7aztmxuph3HAZBVZgVtHH9+bHkruXsYF3MOw6DoCpc+Wf8xlJL0P1LY3UZ2HEYBFXh/J24+nRu9oYQFPw4fSeuPp33rzeOQFCY4FzQr1O3Wi/vz7lcg8IEE4IevLq7aqUXD+OMC3p45P7ClHFQGGX4GtTQc2ImCYYZ7cWfHIeg4MfgOKhlbPS6FNNelh2HQVAVxmaSLq9craUY9rLsOAyCqjB0Br24LjCXYtrLsuMwCKrCyDWobXq+4wjTXpYdh0FQFQZ68VdvHoKCH/Zx0JGrVgSFQcyCDs1YIigMYhV0bMYSQWEQ4zXo4IwlgsIgtl581/uGoOCHaRx02DMEhUEsffLOdw1BwQ+DoL1vGoKCH/2CzuiDoDBIt6D9bxmCgh+9ghreMQQFPzoFtbxhCAp9fwe2K1LXDqZSEBS+cUlil6C2khAUvgkT1FgQgsI3UYJay0FQ+CZIUHMxCArfxF2DTodE0CcSIujAMAGCwjcOSbwcqLJfgSIo/GU+iZe39bJ34b8QFLwG6i9vjGgfA/056vqlvgMndxwGQedwEvTr3EH7JNK/wy5f6jtwcsdhENSB6SReCGqfhf993NVLfQdO7jgMgjqwWFD7Mqb3Ay9e6jtwcsdhENSBeUHPrkG3j9+GsF0v9R04ueMwCOrA0l78tnvQH7Xrpb4DJ3ccBkEdWDkOujUedQfteqnvwMkdh0FQBxbOJG1XOxhDIugTWSfodrWDNSSCPpFlgm5XO5hDIugTWSXodrXDbEjDgZM7DoOgDiwSdLvawR4SQR/JGkE/X0BQGGSJoJfC2kN2B0HQWqwQ9PIT3x6yPwiC1mKBoMNuuQRB0Fr4C3o55mQPaQmCoLVwF/Ry0N4e0hQEQWvhLejlrKc9pC0IgtbCWdCDcAgKg/gKehQNQWEQV0EPgyEoDOIp6HGsWUEN3+9D0Fo4CnoSyuUM6nwggqbAT9CzSKZS5r4NjaC1cBP0sH/kdh/n/rr47jgMgjrgJeh5nMC3CkFr4fLhe3l3MASFQXwGgK6iICgMEjNCiaAwiIegjoOT8yBoLRwE7bg/LYLCIPOCXnbhh0oZBkFrMS3o2RjoRCnDIGgtDtZvnoyub/tnp5NIDNTDBPYJ9G3/ROmdQNBazAl6uYwpHgStxZSgl+tAbwBBazEj6LZ/6X4QtBYTgjZMFQBBazEuaOOjXgEErcWwoI2+kgQIWotRQRuDTRogaC0GBW2M1ouAoLUYE7Qx3akCgtZiSNDGehEZELQWI4I2FtzpgKC1GBD0av3IvSBoLSYXi1wGCQdBa2EWtLVqTumdQNBaWAVtfr9D6Z1A0FoYBW104c+DhIOgtbAJ2hgDvQgSDoLWwiRoYxLpKkg4CFoLi6CNWfjLIOEgaC0MgjaWMV0HCQdBa9EvaGMdaEeQcBC0Ft2CbgcvB3+r+BIErUWvoFv7ZT2CBP2Vj8P/neKZykGnoFv7ZUFiBN16/sOKZyoHfYIeuKpIiKBb1yeKeKZy0CXowYe9JBGCbu/JObr+Fs9UDnoEPegtaRIg6Pt82vY5vabWbcxNh6AHw02ihAt6HEw8Uzm4FvRgvF6VCEH3p0cEXcWloAcTnrKEDjO9zqUIuoorQQ9WjOgSKeif8+jh1aZ4pnJwIaj4vGaD0Jmk8xjimcrBuaCtreJpZ6qzFqeCZryyQtBanAma8soKQWvRvZqp5wgFELQWJzPJCDqJeKZycJjEw78fJ552BK3FURIPuvAnR4iAoLU4sRBBJxHPVA5OeuoIOol4pnJwMtSJoJOIZyoHJ3NFCDqJeKZycDLZjqCTiGcqByerlRB0EvFM5eBwuWfWL9MiaC0+k/jPz/+H6lnNNIN4pjKwO03+8vN7rJ71oBOIZyoBu9Pk2/Vnc7Be/tuKCFqH3WnyX0aPBD3+5FcBQevwaeF2vOnn5eNPfhUQtA4fFv7O54GJh5/8OiBoHd51e09n+7McQS0IZykHb6fJ/dVmozeEoBaEs5SEX6fJvvEkrkEtCGcpC6/TZO+IPL14A8ppSkPr+vNt0+eLjIN2I52nLByeP1ksMo14pnJw0uVB0EnEM5WD7TiPCDqJeKZycPjtYgSdRjxTOTjp8SDoJOKZysFJEhF0EvFMpeBsyAhBJxHPVAZOrkARdBrxTCXgpAv/haDTiGdKn4tlHwg6iXim5DmZRPo62SSedgStwsks/NfZJvG0I2gRtrdfJ3t0vqwCgtZg+/h9skvfyyogaAl+zp9ni+cQdBLxTCnz8vNs+TGCTiKeKWH++Xn2BQ4EnUQ8U7r8vv5kPeg6xDMly/b7AYKuQzxTqmxvjxB0HeKZEuVX1rgGXYt4pjR5Sxq9+KWIZ0qSj5wxDroS8Uwp8pkyBF2JeKYE2fvJR/xCxDOlR8NPOkkLEc+UHLt8Mcy0FvFMqbFPF4KuRTxTYjSyhaBrEc+UFq1kcQ26FvFMSdHOFb34pYhnSolDBxkHbW/7IaYqj+ckU/ZN4ml3O4Nu000Vz5QOp+cK8ybxtHsJupmCzVXl4ZzmCUHPNiNoAOdpQtCzzQi6np73wbRJPO1+16DTfy1CPFMadHQFjJvE084wUyouc4Sgh9tP703pWpXncp0iBD3cPD3OJJ4pAToyhKCHmxF0NT0JQtDDzQi6mK78HE3SH0/2iacdQbPQlx57EsXTzjBTEjqzg6DrEM/UvfQmB0E9Yi05vDQLzyPiaXdtOdegi+hPDYKebUbQNRgyg6BnmxF0CZbEIGh7o8OCevVM3YYpLwg6G2dpiIrY0oKg6xDP1E0Ys4KgJ9v5iPfHmhQEPdtMJ8kbc04Q9GwzgjoToZt42hFUmIGM2Lr8Pjc0WAqC6jKSkHJJZDWTLEP5KJdEhplUGUtHuSSyWESUwWyUSyKCajKajHJJdJqL/+Ludq4M56JcEunFKzKeinJJ5CNekIlMlEsivXg9ZhJRLokIKsdUHsol0UVQbgHuCEl8g+V2YpDEd+jFazGbhHJJRFAppnNQLokIqsR8CsolkdVMQjhkoFwSGWbSwSMB5ZLoeAalAzqHS/vLJdHvGnR6nKlcbm34NL9cEhFUBKfWl0sigmrg1fhySURQCdzaXi6JDDMp4Nf0cklkmEkAx5aXS6LnYpFJyuW2F6/+UYb7MJhxneqco1JaLbi2u1wSEfRufJtdLokIejPOrS6XRNdePMNMZrwbXS6JnEFvxb3N5ZLIMNOd+De5XBJ9BHUZ2SiX20sWtLhcEl0E3SyB5qtShRUNLpdEBL2NJe0tl0QEvYs1zS2XRAS9iUWtLZdEH0G5s4iVVY0tl0SGmW5hWVvLJRFB72BdU8slEUFvYGFLyyURQeNZ2dBySUTQcJa2s1wSETSatc0sl0QEDWZxK8slEUFjWd3IcklE0FCWt7FcEhE0EpJoBkEDCWhhuSQiaBwRDSyXRAQNI6R95ZKIoJ1M37cjpnnaSRwAQS1MVDGodQmSaANBLYxXMapxCZJoA0EtDFcxrG0JkmgDQS2MVjGuaQmSaANBLUj3j6KLigFBLYxVMbJhCZJoA0EtDFUxtF0JkmgDQS2MVDG2WQmSaANBLQxUMbhVCZJoA0Et2KsY3agESbSBoBbMVQxvU4Ik2kBQC9YqxjcpQRJtIKgFYxVvaFGCJNpAUAu2Kt7RoARJtIGgFkxVvKU9CZJoA0EtWKp4T3MSJNEGglowVPGm1iRIog0EtdBfxbsakyCJNhDUglC21ApehVDKE+S2t4r3NSVBEm0gqIXOKt7YkgRJtIGgFvqqeGdDEiTRBoJa6Krire1IkEQbCGqhp4r3NiNBEm0gqIWOKt7cigRJtIGgFq6reHcj7i7fHQS1cFnF29twewW8CRL099FHNzdKkNurKt7fhPtr4EyMoL+d3I5iJcjtRRUFWiBQBV9CBN1+Hb0dBkuQ2/Mq3t0/mr3/niQRgr6dM+sKKlF/iUp4EiDo9nUqaKb/92dV1Ki+Ri0cuV9QhxLCOKmiSO1FquFHhKBvp8eigqpUXqUeboQOM/3IGdeLd75+OIwi44VMRbyIFPSvJuHjoH6BjyLpaKFTEydCZ5JuGqVZLqiQFUJV8eEJU52rBVWSQqkuLiDodCQVJxioX0xSQSvZIAiCTkbCz7Ug6Fwk/FwMgk5Fws/VIOhMJPxcDoJORMLP9SDoeCT8DABBhyPhZwQIOhoJP0NA0MFI+BkDgo5Fws8gEHQoEn5GgaAjkfAzDAQdiISfcSCoPRJ+BoKg5kj4GQmCWiPhZygIaoyEn7EgqC0SfgaDoKZI+BkNgt4TCTpBUEsgBA0HQS1xEDQcBLWEQdBwENQSBUHDQVBLEAQNB0EtMRA0HAS1hEDQcBDUEgFBw0FQSwAEDQdBLccjaDgIajkcQcNBUMvRCBoOgvYfXO3mxSlA0O5jt//n4jE0GATtPXT7MxePobEgaO+R269/EMYTBe3+axjb5xMEDeeJgvYWtu2eIWg4CNp7GNegt4Cg3UfRi78DBO0/iHHQG0DQ/mO4/rwBBO0+5Poo8AdBe4+4PgoWUF/Q9pXjaWGHGxE0nPKCHvS9zwo73oag4VQX9Gj08qSwsU2whvKC/vrXV9jYuRUWgaCWWiBoOAhqqQSChlNeUNs16NjwKKyjuqC2XvzY/BIspLyglnHQkQl6WEt9QftnkgZWOMFqENRSPoKGg6CW4hE0HAS1lI6g4SCopXAEDQdBLWUjaDgIaikaQcNBUEvJCBqO0Htzm6D9BSNoOAhqKRdBw0FQS7EIGs7jBTWViqDhPF1QW6EIGs7DBTWWiaDhPFtQa5EIGs6jBTWXiKDhPFlQe4EIGs6DBR0oD0HDea6gI8UhaDiPFXSoNAQN57GCukWCpTxU0ME7JSNoOM8UdOwKFEFv4JGCjnXhW5FgNU8UdGwMtBUJlvNAQccmkVqRYD3PE3RsFr4VCQJ4nKBjy5hakSCCpwk6tg60FQlCeJigYwvpW5EghmcJ2no0FgmCeJSgrVPpWCSI4kmCtq5FxyJBGA8StNWZH4sEcTxH0NZo6FgkCOQxgramk8YiQSRPEbQ1Hz8WCUJ5iKCtBU1jkSCWZwjaWhE6FgmCeYSgrSX1Y5EmawJmHiGoW2EIGs4DBG1+/whBk1Bf0OYVaPPPI/bEgmDKC9rswrf/wGxXMIiluqDNMdA/0g4YiqDhFBe0NYn0khZBE1Bb0NYsPIKmorSgrWVMXwiaisqCttaBfj/nGjQPhQX9Pkm+eNtCLz4NdQU9C8c4aBrKCnoRjZmkJFQV9CoYgiahqKCXsRA0CTUFvQ6FoEkoKWhHJARNQkVBewIhaBIKCtoVB0GTUE/QvjAImoRygnZGQdAkVBO0N4i5sK09awqLKSaoUHPABaF31KGE/hAImoRSghoiIGgSKglqCYCgSSgkqOl4BE1CHUFthyNoEsoIajwaQZNQRVDrwQiahCKC2sfdJwqDQGoIyvczy1JCUL6fWZcKgvL9zMIUEHToOARNQn5BWThXmvSChmoN4WQXNPS6AOJJLmhoxwpuILegoSNTcAepBQ0d2odbyCxo6Nwo3ENiQUMXl8BN5BU0dHUe3EVaQUOXN8NtZBU09PshcB9JBQ39gh3cSE5BQ7+hDHeSUtDQWzzArWQU1KMqCJqEhIK61ARBk5BPUJ+KIGgS0gnqVA8ETUI2Qb2qgaBJSCaoWy0QNAm5BPWrBIImIZWgjnVA0CRkEtSzCgiahESCutYAQZOQR1DfCiBoEtII6lw+giYhi6DexSNoEpII6l46giYhh6D+hSNoEgIEff/7gcd/TfC4hAU2IWgSos6g2+5BfwkrZELQJAQJujUe9ZawxCUETUKMoNvvh0d/L/ighDUqIWgSQgT9reT3Q8M16CKTEDQJ4degu2fnf4d9lUgImoT7BT19cc315+l/CZAiYpjp7ce/Zx0lYNDjiTiD/pyr/vw6PHPtX8VPCJ1JMi4HwU9QnurET1AWFD/hS1hQ/IT/URUUP+EbUUHxE/6gKSh+wl8kBcVP+EFRUPyEF4KC4if8Q09Q/IRfyAmKn/AbNUHxE94QExQ/4R0tQfETPpASFD/hEyVB8RN2SAkKsKNbn5VuyhU7QJqaFq0ogl6QpqZFK4qgF6SpadGKIugFaWpatKIIekGamhatKIJekKamRSuKoBekqWnRiiLoBWlqWrSiCHpBmpoWrWiaZsEzQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBmlBB37+yb/sCfzhvf/xBt5pv91jXzehn1bqrGd+cJLcEf8umcmV/VVS3ko0/59ZZ2fA2pUjn+y3PpO9wuquoJtv+WV9to9v0dl5S/Tz6+B+uLOhHRWUz+l41XUHfaql7a9v3mgkL+lFR4Yy+VU1X0F2Joun8Sino6yVRENSLj+GG1w85Gv12zYp+Iy/o9vZD+H3/n+3fT+le/K+KCmd02/2Q7MX//Hf/80t20O6bXxVNMA4qn9H3t155HPTOYu1QUW+MFU3TLngmCArSIChIg6AgDYKCNAgK0iAoSIOgIA2CgjQICtIgKEiDoCANgoI0CArSIChIg6AgDYKCNAgK0iAoSIOgIA2CgjQICtIgKEiDoCANgoI0CArSIChIg6AgDYKCNAi6Bu7h5gSZWQOCOkFm1oCgTpAZb/7cSnh7v+fxx4t//v3bxttwCJlx5u8t2P+K+PPs75bt969f22Rv3H0/ZMaZTyV/bsz+9xT62mf7eVH4Lx8oQGac+RD0j6Lb740/27e3A6ANmXHm8iP+9Qof8T2QGW9OOkmff8mKTtI1ZAakQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSkQVCQBkFBGgQFaRAUpEFQkAZBQRoEBWkQFKRBUJAGQUEaBAVpEBSk+Q/IGM93oBkn6AAAAABJRU5ErkJggg==" /><!-- -->
Fig. 5. Leave-one-out crossvalidated prediction of the neonatal weight
for 13 odd-toed ungulate species.</p>
<hr />
<p>From the present cross-validation, we found that the (<span class="math inline">\(\log_{10}\)</span>) neonatal body mass can be
predicted with a cross-validated <span class="math inline">\(R^{2}\)</span> of 0.96 (Fig. 5).</p>
</div>
<div id="other-utility-functions" class="section level1">
<h1>Other utility functions</h1>
<div id="influence-matrix" class="section level2">
<h2>Influence matrix</h2>
<p>The influence matrix is used internally to calculate
<strong>PEM</strong>. It is a matrix having as many rows as the number
of vertices (species + nodes) and as many columns as the number of
edges. Any given element of the influence matrix is coding whether a
vertex, which is represented a row of the matrix is influenced an edge,
which is represented by a column of the matrix. In the context of
<strong>PEM</strong>, a vertex is influenced by an edge when the former
has ancestors on the latter or, in other words, when an edge is on the
path leading from a tip to the root of the tree. The influence matrix is
obtained as follows:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="fu">InflMat</span>(perissodactyla.pgraph) <span class="ot">-&gt;</span> res</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>res</span></code></pre></div>
<pre><code>##                      E1 E2 E3 E4 E5 E6 E7 E8 E9 E10 E11 E12 E13 E14 E15 E16 E17
## Rhinoceros sondaicus  1  1  1  1  0  0  0  0  0   0   0   0   0   0   0   0   0
## Rhinoceros unicornis  1  1  1  0  1  0  0  0  0   0   0   0   0   0   0   0   0
## Diceros bicornis      1  1  0  0  0  1  1  0  0   0   0   0   0   0   0   0   0
## Ceratotherium simum   1  1  0  0  0  1  0  1  0   0   0   0   0   0   0   0   0
## Tapirus terrestris    1  0  0  0  0  0  0  0  1   1   1   0   0   0   0   0   0
## Tapirus pinchaque     1  0  0  0  0  0  0  0  1   1   0   1   0   0   0   0   0
## Tapirus indicus       1  0  0  0  0  0  0  0  1   0   0   0   1   0   0   0   0
## Equus grevyi          0  0  0  0  0  0  0  0  0   0   0   0   0   1   1   1   1
## Equus burchelli       0  0  0  0  0  0  0  0  0   0   0   0   0   1   1   1   0
## Equus zebra           0  0  0  0  0  0  0  0  0   0   0   0   0   1   1   0   0
## Equus africanus       0  0  0  0  0  0  0  0  0   0   0   0   0   1   0   0   0
## Equus hemionus        0  0  0  0  0  0  0  0  0   0   0   0   0   1   0   0   0
## n1                    0  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0
## n2                    1  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0
## n3                    1  1  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0
## n4                    1  1  1  0  0  0  0  0  0   0   0   0   0   0   0   0   0
## n5                    1  1  0  0  0  1  0  0  0   0   0   0   0   0   0   0   0
## n6                    1  0  0  0  0  0  0  0  1   0   0   0   0   0   0   0   0
## n7                    1  0  0  0  0  0  0  0  1   1   0   0   0   0   0   0   0
## n8                    0  0  0  0  0  0  0  0  0   0   0   0   0   1   0   0   0
## n9                    0  0  0  0  0  0  0  0  0   0   0   0   0   1   1   0   0
## n10                   0  0  0  0  0  0  0  0  0   0   0   0   0   1   1   1   0
##                      E18 E19 E20 E21
## Rhinoceros sondaicus   0   0   0   0
## Rhinoceros unicornis   0   0   0   0
## Diceros bicornis       0   0   0   0
## Ceratotherium simum    0   0   0   0
## Tapirus terrestris     0   0   0   0
## Tapirus pinchaque      0   0   0   0
## Tapirus indicus        0   0   0   0
## Equus grevyi           0   0   0   0
## Equus burchelli        1   0   0   0
## Equus zebra            0   1   0   0
## Equus africanus        0   0   1   0
## Equus hemionus         0   0   0   1
## n1                     0   0   0   0
## n2                     0   0   0   0
## n3                     0   0   0   0
## n4                     0   0   0   0
## n5                     0   0   0   0
## n6                     0   0   0   0
## n7                     0   0   0   0
## n8                     0   0   0   0
## n9                     0   0   0   0
## n10                    0   0   0   0</code></pre>
</div>
<div id="update-and-forced-pem-parameters" class="section level2">
<h2>Update and forced <strong>PEM</strong> parameters</h2>
<p>The calculation of the influence matrix performed by
<code>PEM.build()</code> for a given phylogenetic graph need not be done
every time new weighting function parameters are to be tried. For that
reason, <strong>MPSEM</strong> provides a function called
<code>PEM.updater()</code> that takes a previously calculated
<strong>PEM</strong> object, applies new edge weighting, and
recalculates the phylogenetic eigenvectors:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">PEM.updater</span>(<span class="at">object =</span> perissodactyla.PEM, <span class="at">a =</span> <span class="dv">0</span>, <span class="at">psi =</span> <span class="dv">1</span>) <span class="ot">-&gt;</span> res</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>res</span></code></pre></div>
<pre><code>## A phylogenetic eigenvector map (PEM) for  12  species:
## Rhinoceros sondaicus,Rhinoceros unicornis,Diceros bicornis,Ceratotherium simum,Tapirus terrestris,Tapirus pinchaque,Tapirus indicus,Equus grevyi ... Equus hemionus 
## obtained from the following phylogenetic graph:
## 
## A graph with 21 edges and 22 vertices. 
## Edge labels: E1 E2 E3 E4 E5 E6 E7 E8 E9 E10 E11 E12 E13 E14 E15 E16 E17 E18 E19 E20 E21 
## Vertex labels: Rhinoceros sondaicus Rhinoceros unicornis Diceros bicornis Ceratotherium simum Tapirus terrestris Tapirus pinchaque Tapirus indicus Equus grevyi Equus burchelli Equus zebra Equus africanus Equus hemionus n1 n2 n3 n4 n5 n6 n7 n8 n9 n10 
## Available edge information:  distance 
## Available vertex information:  species</code></pre>
<p>The result of <code>PEM.build()</code> and <code>PEM.updater()</code>
does not contain all the information necessary to predict trait values.
Hence, neither of these functions is given information about the
response variable and auxiliary traits. To perform these preliminary
calculations, <strong>MPSEM</strong> provides the user with function
<code>PEM.forcedSimple()</code> that produce the same output as
<code>PEM.fitSimple()</code> with user-provided values of weighting
parameters. It is called as follows:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">PEM.forcedSimple</span>(</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>  <span class="at">y =</span> perissodactyla.train[,<span class="st">&quot;log.neonatal.wt&quot;</span>],</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a>  <span class="at">x =</span> perissodactyla.train[,<span class="st">&quot;log.female.wt&quot;</span>],</span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>  <span class="at">w =</span> perissodactyla.pgraph,</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a>  <span class="at">a =</span> steepness,</span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a>  <span class="at">psi =</span> evol_rate</span>
<span id="cb32-7"><a href="#cb32-7" tabindex="-1"></a>) <span class="ot">-&gt;</span> res</span>
<span id="cb32-8"><a href="#cb32-8" tabindex="-1"></a>res</span></code></pre></div>
<pre><code>## A phylogenetic eigenvector map (PEM) for  12  species:
## Rhinoceros sondaicus,Rhinoceros unicornis,Diceros bicornis,Ceratotherium simum,Tapirus terrestris,Tapirus pinchaque,Tapirus indicus,Equus grevyi ... Equus hemionus 
## obtained from the following phylogenetic graph:
## 
## A graph with 21 edges and 22 vertices. 
## Edge labels: E1 E2 E3 E4 E5 E6 E7 E8 E9 E10 E11 E12 E13 E14 E15 E16 E17 E18 E19 E20 E21 
## Vertex labels: Rhinoceros sondaicus Rhinoceros unicornis Diceros bicornis Ceratotherium simum Tapirus terrestris Tapirus pinchaque Tapirus indicus Equus grevyi Equus burchelli Equus zebra Equus africanus Equus hemionus n1 n2 n3 n4 n5 n6 n7 n8 n9 n10 
## Available edge information:  distance, a, psi 
## Available vertex information:  species</code></pre>
<p>It is noteworthy that contrary to function
<code>PEM.fitSimple()</code>, function <code>PEM.forcedSimple()</code>
can be used to apply different weighting parameters for different
edges.</p>
</div>
<div id="pem-scores" class="section level2">
<h2><strong>PEM</strong> scores</h2>
<p><strong>PEM</strong> scores are the values of target species on the
eigenfunctions underlying the <strong>PEM</strong>. These scores are
calculated from the graph locations and a <strong>PEM</strong> object
using function <code>Locations2PEMscores()</code> as follows:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">Locations2PEMscores</span>(</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>  <span class="at">object =</span> perissodactyla.PEM_opt2,</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>  <span class="at">gsc =</span> perissodactyla.loc</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>) <span class="ot">-&gt;</span> scores</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>scores</span></code></pre></div>
<pre><code>## $scores
##                                V_1      V_2         V_3           V_4       V_5
## Dicerorhinus sumatrensis 0.2439632 0.287517 -0.01215849 -2.081668e-17 -0.351012
##                                    V_6          V_7          V_8          V_9
## Dicerorhinus sumatrensis -0.0007381739 -0.005777055 5.085388e-17 1.995388e-16
##                                   V_10         V_11
## Dicerorhinus sumatrensis -3.768614e-17 4.815588e-16
## 
## $VarianceFactor
##   Dicerorhinus sumatrensis
## y              0.004808351</code></pre>
<p>The function is used internally by <code>PEM-class</code>
<code>predict</code> method, and therefore need not be called when
performing linear phylogenetic modelling as exemplified above. It comes
in handy when the <strong>PEM</strong> is used together with other
modelling approaches (e.g. multivariate regression trees, linear
discriminant analysis, artificial neural networks) that have
<code>predict</code> methods that are not specially adapted for
phylogenetic modelling.</p>
</div>
<div id="miscellaneous" class="section level2">
<h2>Miscellaneous</h2>
<p>Package <strong>MPSEM</strong> comes with functions, some implemented
in the <strong>C</strong> language, to simulate quantitative traits
evolution by Ornstein-Uhlenbeck process on potentially large phylogenies
<span class="citation">(Butler and King 2004)</span>. These functions
are only useful to perform simulations, which is a rather advanced
matter outside the scope of the present tutorial. I refer the user to
<strong>MPSEM</strong> help files for further details.</p>
<p>In addition to function <code>Phylo2DirectedGraph()</code>, which we
have seen previously <strong>MPSEM</strong> also has built-in graph
manipulation functions to populate a graph with vertices, add and remove
vertices and edges, etc. These functions were mainly intended to be
called internally by <strong>MPSEM</strong> functions. They were made
visible upon loading the package because of their potential usefulness
to some advanced applications that are outside the scope of the present
tutorial. Again, I refer the user to <strong>MPSEM</strong> help files
for further details.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Borcard2002" class="csl-entry">
Borcard, D., and P. Legendre. 2002. <span>“All-Scale Spatial Analysis of
Ecological Data by Means of Principal Coordinates of Neighbour
Matrices.”</span> <em>Ecological Modelling</em> 153: 51–68. <a href="https://doi.org/10.1016/S0304-3800(01)00501-4">https://doi.org/10.1016/S0304-3800(01)00501-4</a>.
</div>
<div id="ref-Butler2004" class="csl-entry">
Butler, M. A., and A. A. King. 2004. <span>“Phylogenetic Comparative
Analysis: A Modeling Approach for Adaptive Evolution.”</span>
<em>American Naturalist</em> 164: 683–95. <a href="https://doi.org/10.1086/426002">https://doi.org/10.1086/426002</a>.
</div>
<div id="ref-Diniz2012" class="csl-entry">
Diniz-Filho, J. A. F., L. M. Bini, T. F. Rangel, I. Morales-Castilla, M.
A. Olalla-Tarraga, M. A. Rodrıguez, and B. A. Hawkins. 2012. <span>“On
the Selection of Phylogenetic Eigenvectors for Ecological
Analyses.”</span> <em>Ecography</em> 34: 239–49. <a href="https://doi.org/10.1111/j.1600-0587.2011.06949.x">https://doi.org/10.1111/j.1600-0587.2011.06949.x</a>.
</div>
<div id="ref-Dray2006" class="csl-entry">
Dray, S., P. Legendre, and P. R. Peres-Neto. 2006. <span>“Spatial
Modelling: A Comprehensive Framework for Principal Coordinate Analysis
of Neighbour Matrices (PCNM).”</span> <em>Ecological Modelling</em>,
483–93. <a href="https://doi.org/10.1016/j.ecolmodel.2006.02.015">https://doi.org/10.1016/j.ecolmodel.2006.02.015</a>.
</div>
<div id="ref-Guenard2013" class="csl-entry">
Guénard, G., P. Legendre, and P. Peres-Neto. 2013. <span>“Phylogenetic
Eigenvector Maps: A Framework to Model and Predict Species
Traits.”</span> <em>Methods in Ecology and Evolution</em> 4: 1120–31. <a href="https://doi.org/10.1111/2041-210X.12111">https://doi.org/10.1111/2041-210X.12111</a>.
</div>
<div id="ref-Guenard2014" class="csl-entry">
Guénard, G., P. C. von der Ohe, S. C. Walker, S. Lek, and P. Legendre.
2014. <span>“Using Phylogenetic Information and Chemical Properties to
Predict Species Tolerances to Pesticides.”</span> <em>Proceedings of the
Royal Society B: Biological Sciences</em> 281 (20133239): 20133239. <a href="https://doi.org/10.1098/rspb.2013.3239">https://doi.org/10.1098/rspb.2013.3239</a>.
</div>
<div id="ref-Guenard2011" class="csl-entry">
Guénard, G., P. C. von der Ohe, D. de Zwart, P. Legendre, and S. Lek.
2011. <span>“Using Phylogenetic Information to Predict Species
Tolerances to Toxic Chemicals.”</span> <em>Ecological Applications</em>
21: 3178–90. <a href="https://doi.org/10.1890/10-2242.1">https://doi.org/10.1890/10-2242.1</a>.
</div>
<div id="ref-Hurvich1993" class="csl-entry">
Hurvich, C. M., and C.-L. Tsai. 1993. <span>“A Corrected
<span>A</span>kaike Information Criterion for Vector Autoregressive
Model Selection.”</span> <em>Journal of Time Series Analysis</em> 14:
271–79. <a href="https://doi.org/10.1111/j.1467-9892.1993.tb00144.x">https://doi.org/10.1111/j.1467-9892.1993.tb00144.x</a>.
</div>
<div id="ref-Moran1948" class="csl-entry">
Moran, P. A. P. 1948. <span>“The Interpretation of Statistical
Maps.”</span> <em>Journal of the Royal Statistical Society: Series B
(Methodological)</em> 10: 243–51. <a href="https://doi.org/10.1111/j.2517-6161.1948.tb00012.x">https://doi.org/10.1111/j.2517-6161.1948.tb00012.x</a>.
</div>
<div id="ref-Moran1950" class="csl-entry">
———. 1950. <span>“Notes on Continuous Stochastic Phenomena.”</span>
<em>Biometrika</em> 37: 17–23. <a href="https://doi.org/10.1093/biomet/37.1-2.17">https://doi.org/10.1093/biomet/37.1-2.17</a>.
</div>
<div id="ref-Purvis1995" class="csl-entry">
Purvis, A., and A. Rambaut. 1995. <span>“Comparative Analysis by
Independent Contrasts (<span>CAIC</span>): An <span>A</span>pple
<span>M</span>acintosh Application for Analysing Comparative
Data.”</span> <em>Computer Applications in the Biosciences</em> 11:
247–51. <a href="https://doi.org/10.1093/bioinformatics/11.3.247">https://doi.org/10.1093/bioinformatics/11.3.247</a>.
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
